// @web3auth/modal/dist/lib.esm/packages/modal/src/ui/constants.js
const PAGES = {
  LOGIN: "login",
  CONNECT_WALLET: "connect_wallet"};
const CONNECT_WALLET_PAGES = {
  CONNECT_WALLET: "Connect Your Wallet",
  SELECTED_WALLET: "Selected Wallet"
};
const DEFAULT_LOGO_LIGHT = "https://images.web3auth.io/web3auth-logo-w.svg"; // logo used on light mode
const DEFAULT_LOGO_DARK = "https://images.web3auth.io/web3auth-logo-w-light.svg"; // logo used on dark mode

const WALLET_CONNECT_LOGO = "https://images.web3auth.io/login-wallet-connect.svg";
const DEFAULT_PRIMARY_COLOR = "#0364FF";
const DEFAULT_ON_PRIMARY_COLOR = "#FFFFFF";
const DEFAULT_METAMASK_WALLET_REGISTRY_ITEM = {
  app: {
    android: "io.metamask",
    chrome: "nkbihfbeogaeaoehlefnkodbefgpgknn",
    edge: "ejbalbakoplchlghecdalmeeeajnimhm",
    firefox: "ether-metamask",
    ios: "id1438144202"
  },
  chains: ["eip155:1"],
  imgExtension: "webp",
  injected: [{
    injected_id: "isMetaMask",
    namespace: "eip155"
  }],
  mobile: {
    inAppBrowser: "https://metamask.app.link/dapp",
    native: "metamask://wc",
    universal: "https://metamask.app.link/wc"
  },
  name: "MetaMask",
  primaryColor: "#E2761B",
  walletConnect: {
    sdks: ["sign_v1", "sign_v2"]
  }
};

export { CONNECT_WALLET_PAGES, DEFAULT_LOGO_DARK, DEFAULT_LOGO_LIGHT, DEFAULT_METAMASK_WALLET_REGISTRY_ITEM, DEFAULT_ON_PRIMARY_COLOR, DEFAULT_PRIMARY_COLOR, PAGES, WALLET_CONNECT_LOGO };
