// @toruslabs/customauth/dist/lib.cjs/index.js
'use strict';

var HandlerFactory = require('./handlers/HandlerFactory.js');
var login = require('./login.js');
var enums = require('./utils/enums.js');
var error = require('./utils/error.js');
var helpers = require('./utils/helpers.js');
var interfaces = require('./utils/interfaces.js');
var PopupHandler = require('./utils/PopupHandler.js');
var StorageHelper = require('./utils/StorageHelper.js');



exports.createHandler = HandlerFactory.createHandler;
exports.CustomAuth = login.CustomAuth;
exports.AUTH_CONNECTION = enums.AUTH_CONNECTION;
exports.REDIRECT_PARAMS_STORAGE_METHOD = enums.REDIRECT_PARAMS_STORAGE_METHOD;
exports.SENTRY_TXNS = enums.SENTRY_TXNS;
exports.UX_MODE = enums.UX_MODE;
exports.serializeError = error.serializeError;
exports.broadcastChannelOptions = helpers.broadcastChannelOptions;
exports.constructURL = helpers.constructURL;
exports.decodeToken = helpers.decodeToken;
exports.eventToPromise = helpers.eventToPromise;
exports.getPopupFeatures = helpers.getPopupFeatures;
exports.getTimeout = helpers.getTimeout;
exports.getUserId = helpers.getUserId;
exports.handleRedirectParameters = helpers.handleRedirectParameters;
exports.isFirefox = helpers.isFirefox;
exports.isMobileOrTablet = helpers.isMobileOrTablet;
exports.loginToConnectionMap = helpers.loginToConnectionMap;
exports.objectToAuthDataMap = helpers.objectToAuthDataMap;
exports.padUrlString = helpers.padUrlString;
exports.randomId = helpers.randomId;
exports.storageAvailable = helpers.storageAvailable;
exports.validateAndConstructUrl = helpers.validateAndConstructUrl;
exports.EMAIL_FLOW = interfaces.EMAIL_FLOW;
exports.PopupHandler = PopupHandler.PopupHandler;
exports.StorageHelper = StorageHelper.StorageHelper;
