// @web3auth/no-modal/dist/lib.cjs/connectors/wallet-connect-v2-connector/index.js
'use strict';

var config = require('./config.js');
var walletConnectV2Connector = require('./walletConnectV2Connector.js');
var WalletConnectV2Provider = require('./WalletConnectV2Provider.js');



Object.defineProperty(exports, "DEFAULT_EIP155_METHODS", {
	enumerable: true,
	get: function () { return config.DEFAULT_EIP155_METHODS; }
});
Object.defineProperty(exports, "DEFAULT_EIP_155_EVENTS", {
	enumerable: true,
	get: function () { return config.DEFAULT_EIP_155_EVENTS; }
});
Object.defineProperty(exports, "DEFAULT_SOLANA_EVENTS", {
	enumerable: true,
	get: function () { return config.DEFAULT_SOLANA_EVENTS; }
});
Object.defineProperty(exports, "DEFAULT_SOLANA_METHODS", {
	enumerable: true,
	get: function () { return config.DEFAULT_SOLANA_METHODS; }
});
exports.getNamespacesFromChains = config.getNamespacesFromChains;
exports.getRequiredNamespaces = config.getRequiredNamespaces;
exports.getSupportedEventsByNamespace = config.getSupportedEventsByNamespace;
exports.getSupportedMethodsByNamespace = config.getSupportedMethodsByNamespace;
exports.getWalletConnectV2Settings = config.getWalletConnectV2Settings;
exports.walletConnectV2Connector = walletConnectV2Connector.walletConnectV2Connector;
exports.WalletConnectV2Provider = WalletConnectV2Provider.WalletConnectV2Provider;
