// @web3auth/modal/dist/lib.esm/packages/modal/src/ui/components/ConnectWallet/ConnectWalletChainFilter/ConnectWalletChainFilter.js
import { CHAIN_NAMESPACES } from '@web3auth/no-modal';
import { useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import i18nInstance from '../../../localeImport.js';
import { cn, getIcons } from '../../../utils.js';
import { jsx, jsxs } from 'react/jsx-runtime';

function ConnectWalletChainFilter(props) {
  const {
    isDark,
    isLoading,
    selectedChain,
    setSelectedChain,
    chainNamespace
  } = props;
  const [t] = useTranslation(undefined, {
    i18n: i18nInstance
  });
  const chains = useMemo(() => {
    const chains = [{
      id: "all",
      name: "modal.allChains",
      icon: ""
    }];
    for (const chain of chainNamespace) {
      if (chain === CHAIN_NAMESPACES.EIP155 || chain === CHAIN_NAMESPACES.SOLANA) {
        chains.push({
          id: chain,
          name: chain === CHAIN_NAMESPACES.EIP155 ? "EVM" : chain,
          icon: chain === CHAIN_NAMESPACES.EIP155 ? "ethereum" : chain
        });
      }
    }
    return chains;
  }, [chainNamespace]);
  if (isLoading) {
    return /*#__PURE__*/jsx("div", {
      className: "w3a--flex w3a--items-center w3a--justify-between w3a--gap-x-2",
      children: Array.from({
        length: chains.length
      }).map((_, index) => /*#__PURE__*/jsx("div", {
        className: "w3a--h-12 w3a--w-[100px] w3a--animate-pulse w3a--rounded-2xl w3a--bg-app-gray-200 dark:w3a--bg-app-gray-700"
      }, `chain-loader-${index}`))
    });
  }
  return /*#__PURE__*/jsx("div", {
    className: "w3a--flex w3a--items-center w3a--justify-items-start w3a--gap-x-4",
    children: chains.map(chain => /*#__PURE__*/jsxs("button", {
      type: "button",
      className: cn("w3a--flex w3a--w-[104px] w3a--items-center w3a--justify-center w3a--gap-x-1 w3a--text-xs w3a--font-medium w3a--px-4 w3a--py-3 w3a--text-app-gray-500 dark:w3a--text-app-gray-300 hover:w3a--bg-app-gray-200 dark:hover:w3a--bg-app-gray-700 w3a--h-12 w3a--rounded-2xl w3a--border w3a--border-transparent", {
        "w3a--bg-app-gray-100 dark:w3a--bg-app-gray-800 w3a--border-app-gray-200 dark:w3a--border-app-gray-700 w3a--text-app-gray-900 dark:w3a--text-app-white hover:!w3a--bg-app-gray-100 dark:hover:!w3a--bg-app-gray-800": selectedChain === chain.id
      }),
      onClick: () => setSelectedChain(chain.id),
      children: [chain.icon && /*#__PURE__*/jsx("img", {
        src: getIcons(isDark ? `${chain.icon}-dark` : `${chain.icon}-light`),
        alt: chain.name,
        className: "w3a--size-5 w3a--object-contain"
      }), /*#__PURE__*/jsx("span", {
        className: "first-letter:w3a--capitalize",
        children: t(chain.name)
      })]
    }, chain.id))
  });
}

export { ConnectWalletChainFilter as default };
