// @toruslabs/ethereum-controllers/dist/lib.esm/AccountAbstraction/smartAccounts/NexusSmartAccount.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { toNexusSmartAccount } from 'permissionless/accounts';
import { entryPoint07Address } from 'viem/account-abstraction';
import { SMART_ACCOUNT } from '../../utils/constants.js';

class NexusSmartAccount {
  constructor(options) {
    _defineProperty(this, "name", SMART_ACCOUNT.NEXUS);
    _defineProperty(this, "options", void 0);
    this.options = options;
  }
  async getSmartAccount(params) {
    var _this$options, _this$options2, _this$options3;
    return toNexusSmartAccount(_objectSpread(_objectSpread(_objectSpread({}, this.options || {}), {}, {
      entryPoint: {
        address: ((_this$options = this.options) === null || _this$options === void 0 || (_this$options = _this$options.entryPoint) === null || _this$options === void 0 ? void 0 : _this$options.address) || entryPoint07Address,
        version: ((_this$options2 = this.options) === null || _this$options2 === void 0 || (_this$options2 = _this$options2.entryPoint) === null || _this$options2 === void 0 ? void 0 : _this$options2.version) || "0.7"
      },
      version: ((_this$options3 = this.options) === null || _this$options3 === void 0 ? void 0 : _this$options3.version) || "1.0.0"
    }, params), {}, {
      owners: [params.walletClient],
      client: params.client
    }));
  }
}

export { NexusSmartAccount };
