// @toruslabs/ethereum-controllers/dist/lib.esm/AccountAbstraction/smartAccounts/TrustSmartAccount.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { toTrustSmartAccount } from 'permissionless/accounts';
import { entryPoint06Address } from 'viem/account-abstraction';
import { SMART_ACCOUNT } from '../../utils/constants.js';

class TrustSmartAccount {
  constructor(options) {
    _defineProperty(this, "name", SMART_ACCOUNT.TRUST);
    _defineProperty(this, "options", void 0);
    this.options = options;
  }
  async getSmartAccount(params) {
    var _this$options, _this$options2;
    return toTrustSmartAccount(_objectSpread(_objectSpread(_objectSpread({}, this.options || {}), {}, {
      entryPoint: {
        address: ((_this$options = this.options) === null || _this$options === void 0 || (_this$options = _this$options.entryPoint) === null || _this$options === void 0 ? void 0 : _this$options.address) || entryPoint06Address,
        version: ((_this$options2 = this.options) === null || _this$options2 === void 0 || (_this$options2 = _this$options2.entryPoint) === null || _this$options2 === void 0 ? void 0 : _this$options2.version) || "0.6"
      }
    }, params), {}, {
      owner: params.walletClient,
      client: params.client
    }));
  }
}

export { TrustSmartAccount };
