// @web3auth/no-modal/dist/lib.cjs/vue/Web3AuthProvider.js
'use strict';

var vue = require('vue');
var analytics = require('../base/analytics.js');
require('@toruslabs/base-controllers');
var index = require('../base/composables/index.js');
require('@babel/runtime/helpers/defineProperty');
require('@web3auth/auth');
var index$1 = require('../base/errors/index.js');
require('../base/wallet/index.js');
require('../base/connector/connectorStatus.js');
var constants = require('../base/connector/constants.js');
require('jwt-decode');
var loglevel = require('../base/loglevel.js');
require('../base/plugin/errors.js');
require('../base/plugin/IPlugin.js');
require('@toruslabs/constants');
require('@toruslabs/http-helpers');
var noModal = require('../noModal.js');
var WalletServicesInnerProvider = require('./WalletServicesInnerProvider.js');

const Web3AuthProvider = vue.defineComponent({
  name: "Web3AuthProvider",
  props: {
    config: {
      type: Object,
      required: true
    }
  },
  setup(props) {
    const web3Auth = vue.shallowRef(null);
    const provider = vue.ref(null);
    const isMFAEnabled = vue.ref(false);
    const status = vue.ref(null);
    const isInitializing = vue.ref(false);
    const initError = vue.ref(null);
    const isInitialized = vue.ref(false);
    const isConnected = vue.ref(false);
    const isAuthorized = vue.ref(false);
    const getPlugin = name => {
      if (!web3Auth.value) throw index$1.WalletInitializationError.notReady();
      return web3Auth.value.getPlugin(name);
    };
    const setIsMFAEnabled = isMfaEnabled => {
      isMFAEnabled.value = isMfaEnabled;
    };
    vue.watch(() => props.config, (newConfig, _, onInvalidate) => {
      const resetHookState = () => {
        provider.value = null;
        isMFAEnabled.value = false;
        isConnected.value = false;
        isAuthorized.value = false;
        status.value = null;
      };
      onInvalidate(() => {
        if (web3Auth.value) {
          web3Auth.value.cleanup();
        }
      });
      resetHookState();
      const {
        web3AuthOptions
      } = newConfig;
      const web3AuthInstance = new noModal.Web3AuthNoModal(web3AuthOptions);
      web3AuthInstance.setAnalyticsProperties({
        integration_type: analytics.ANALYTICS_INTEGRATION_TYPE.VUE_COMPOSABLES
      });
      web3Auth.value = web3AuthInstance;
    }, {
      immediate: true
    });
    vue.watch(web3Auth, async (newWeb3Auth, _, onInvalidate) => {
      const controller = new AbortController();
      // Invalidate the controller here before calling any async methods.
      onInvalidate(() => {
        controller.abort();
      });
      if (newWeb3Auth) {
        try {
          initError.value = null;
          isInitializing.value = true;
          await newWeb3Auth.init({
            signal: controller.signal
          });
        } catch (error) {
          loglevel.log.error("Error initializing web3auth", error);
          initError.value = error;
        } finally {
          isInitializing.value = false;
        }
      }
    }, {
      immediate: true
    });
    vue.watch(web3Auth, (newWeb3Auth, prevWeb3Auth) => {
      const notReadyListener = () => {
        status.value = web3Auth.value.status;
      };
      const readyListener = () => {
        status.value = web3Auth.value.status;
        isInitialized.value = true;
      };
      const connectedListener = () => {
        status.value = web3Auth.value.status;
        // we do this because of rehydration issues. status connected is fired first but web3auth sdk is not ready yet.
        if (web3Auth.value.status === constants.CONNECTOR_STATUS.CONNECTED) {
          if (!isInitialized.value) isInitialized.value = true;
          isConnected.value = true;
          provider.value = newWeb3Auth.provider;
        }
      };
      const authorizedListener = () => {
        status.value = web3Auth.value.status;
        if (web3Auth.value.status === constants.CONNECTOR_STATUS.AUTHORIZED) {
          isAuthorized.value = true;
          isConnected.value = true;
        }
      };
      const disconnectedListener = () => {
        status.value = web3Auth.value.status;
        isConnected.value = false;
        provider.value = null;
        isMFAEnabled.value = false;
        isAuthorized.value = false;
      };
      const connectingListener = () => {
        status.value = web3Auth.value.status;
      };
      const errorListener = () => {
        status.value = web3Auth.value.status;
      };
      const mfaEnabledListener = () => {
        isMFAEnabled.value = true;
      };
      // unregister previous listeners
      if (prevWeb3Auth && newWeb3Auth !== prevWeb3Auth) {
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.NOT_READY, notReadyListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.READY, readyListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.CONNECTED, connectedListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.AUTHORIZED, authorizedListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.DISCONNECTED, disconnectedListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.CONNECTING, connectingListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.ERRORED, errorListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.REHYDRATION_ERROR, errorListener);
        prevWeb3Auth.removeListener(constants.CONNECTOR_EVENTS.MFA_ENABLED, mfaEnabledListener);
      }
      if (newWeb3Auth && newWeb3Auth !== prevWeb3Auth) {
        status.value = newWeb3Auth.status;
        // web3Auth is initialized here.
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.NOT_READY, notReadyListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.READY, readyListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.CONNECTED, connectedListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.AUTHORIZED, authorizedListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.DISCONNECTED, disconnectedListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.CONNECTING, connectingListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.ERRORED, errorListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.REHYDRATION_ERROR, errorListener);
        newWeb3Auth.on(constants.CONNECTOR_EVENTS.MFA_ENABLED, mfaEnabledListener);
      }
    }, {
      immediate: true
    });
    vue.provide(index.Web3AuthContextKey, {
      web3Auth,
      isConnected,
      isInitialized,
      provider,
      status,
      isInitializing,
      initError,
      isMFAEnabled,
      getPlugin,
      setIsMFAEnabled,
      isAuthorized
    });
  },
  render() {
    var _this$$slots$default;
    return vue.h(WalletServicesInnerProvider.WalletServicesInnerProvider, {}, (_this$$slots$default = this.$slots.default) !== null && _this$$slots$default !== void 0 ? _this$$slots$default : "");
  }
});

exports.Web3AuthProvider = Web3AuthProvider;
