// @toruslabs/base-controllers/dist/lib.cjs/Popup/BroadcastChannelHandler.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var broadcastChannel = require('@toruslabs/broadcast-channel');
var enums = require('../enums.js');
var utils = require('../utils/utils.js');

class BroadcastChannelHandler {
  constructor(channelPrefix, instanceId) {
    _defineProperty(this, "bc", void 0);
    _defineProperty(this, "channel", void 0);
    const queryParameters = new URLSearchParams(window.location.search);
    const windowId = queryParameters.get("windowId");
    this.channel = `${channelPrefix}_${instanceId}_${windowId}`;
    this.bc = new broadcastChannel.BroadcastChannel(this.channel, utils.broadcastChannelOptions);
  }
  getMessageFromChannel() {
    return new Promise((resolve, reject) => {
      this.bc.addEventListener("message", async ev => {
        this.bc.close();
        if (ev.error) {
          reject(ev.error);
        } else {
          resolve(ev.data);
        }
      });
      this.bc.postMessage({
        data: {
          type: enums.POPUP_LOADED
        }
      });
    });
  }
}

exports.BroadcastChannelHandler = BroadcastChannelHandler;
