// @web3auth/no-modal/dist/lib.cjs/base/connector/constants.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');

const CONNECTOR_STATUS = {
  NOT_READY: "not_ready",
  READY: "ready",
  CONNECTING: "connecting",
  CONNECTED: "connected",
  DISCONNECTING: "disconnecting",
  DISCONNECTED: "disconnected",
  ERRORED: "errored",
  AUTHORIZED: "authorized",
  AUTHORIZING: "authorizing"
};
const CONNECTOR_EVENTS = _objectSpread(_objectSpread({}, CONNECTOR_STATUS), {}, {
  CONNECTOR_DATA_UPDATED: "connector_data_updated",
  CACHE_CLEAR: "cache_clear",
  CONNECTORS_UPDATED: "connectors_updated",
  MFA_ENABLED: "mfa_enabled",
  REHYDRATION_ERROR: "rehydration_error"
});
const CONNECTOR_CATEGORY = {
  EXTERNAL: "external",
  IN_APP: "in_app"
};
const CONNECTOR_INITIAL_AUTHENTICATION_MODE = {
  CONNECT_ONLY: "connect-only",
  CONNECT_AND_SIGN: "connect-and-sign"
};

exports.CONNECTOR_CATEGORY = CONNECTOR_CATEGORY;
exports.CONNECTOR_EVENTS = CONNECTOR_EVENTS;
exports.CONNECTOR_INITIAL_AUTHENTICATION_MODE = CONNECTOR_INITIAL_AUTHENTICATION_MODE;
exports.CONNECTOR_STATUS = CONNECTOR_STATUS;
