// @web3auth/base/dist/lib.esm/index.js
export { getChainConfig, getEvmChainConfig, getSolanaChainConfig, getXrplChainConfig } from './chain/config.js';
export { ADAPTER_NAMESPACES, CHAIN_NAMESPACES } from './chain/IChainInterface.js';
export { Web3AuthContextKey } from './composables/index.js';
export { WalletInitializationError, WalletLoginError, WalletOperationsError, WalletProviderError, Web3AuthError, serializeError } from './errors/index.js';
export { default as log } from './loglevel.js';
export { PROVIDER_EVENTS } from './provider/IProvider.js';
export { fetchProjectConfig, fetchWalletRegistry, isHexStrict, normalizeWalletName, signerHost, storageAvailable } from './utils.js';
export { ADAPTER_NAMES, EVM_ADAPTERS, MULTI_CHAIN_ADAPTERS, SOLANA_ADAPTERS, WALLET_ADAPTERS } from './wallet/index.js';
export { UX_MODE, WEB3AUTH_NETWORK } from '@web3auth/auth';
export { ADAPTER_CATEGORY, ADAPTER_EVENTS, ADAPTER_STATUS, BaseAdapter } from './adapter/IAdapter.js';
export { checkIfTokenIsExpired, clearToken, getSavedToken, saveToken } from './adapter/utils.js';
export { NFTCheckoutPluginError, WalletServicesPluginError } from './plugin/errors.js';
export { EVM_PLUGINS, PLUGIN_EVENTS, PLUGIN_NAMESPACES, PLUGIN_STATUS, SOLANA_PLUGINS, WALLET_PLUGINS } from './plugin/IPlugin.js';
export { cloneDeep } from '@toruslabs/base-controllers';
