// @web3auth/solana-provider/dist/lib.esm/providers/injectedProviders/base/baseInjectedProvider.js
import { JRPCEngine, providerFromEngine } from '@web3auth/auth';
import { WalletLoginError } from '@web3auth/base';
import { BaseProvider } from '@web3auth/base-provider';
import { createConfigMiddleware } from '../../../rpc/JrpcClient.js';
import { createSolanaMiddleware } from '../../../rpc/solanaRpcMiddlewares.js';

class BaseInjectedProvider extends BaseProvider {
  constructor({
    config,
    state
  }) {
    super({
      config,
      state
    });
  }
  async switchChain(_) {
    throw WalletLoginError.unsupportedOperation("Chain switching is not supported by this adapter");
  }
  async setupProvider(injectedProvider) {
    const engine = new JRPCEngine();
    const providerHandlers = this.getProviderHandlers(injectedProvider);
    const solanaMiddleware = createSolanaMiddleware(providerHandlers);
    engine.push(solanaMiddleware);
    const configMiddleware = createConfigMiddleware(this.config.chainConfig);
    engine.push(configMiddleware);
    const injectedProviderProxy = this.getInjectedProviderProxy(injectedProvider);
    if (injectedProviderProxy) {
      engine.push(injectedProviderProxy);
    }
    const provider = providerFromEngine(engine);
    this.updateProviderEngineProxy(provider);
    await this.lookupNetwork();
  }
  async lookupNetwork() {
    const {
      chainConfig
    } = this.config;
    this.update({
      chainId: chainConfig.chainId
    });
    return chainConfig.chainId || "";
  }
  getInjectedProviderProxy(_) {
    return undefined;
  }
}

export { BaseInjectedProvider };
