// @toruslabs/ethereum-controllers/dist/lib.esm/Message/MessageController.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { AbstractMessageController, MessageStatus, randomId, MESSAGE_EVENTS } from '@toruslabs/base-controllers';
import log from 'loglevel';
import { METHOD_TYPES } from '../utils/constants.js';
import { validateSignMessageData, normalizeMessageData } from './utils.js';

class MessageController extends AbstractMessageController {
  constructor({
    config,
    state,
    signMessage
  }) {
    super({
      config,
      state
    });
    _defineProperty(this, "name", "MessageController");
    _defineProperty(this, "signMessage", void 0);
    this.signMessage = signMessage;
    this.initialize();
  }
  async processSignMessage(messageId) {
    const msgObject = this.getMessage(messageId);
    if (!msgObject) {
      throw new Error(`Message not found`);
    }
    try {
      const cleanMsgParams = await this.approveMessage(messageId, msgObject.messageParams);
      const rawSig = await this.signMessage(cleanMsgParams.data, cleanMsgParams.from);
      this.updateMessage(_objectSpread(_objectSpread({}, msgObject), {}, {
        rawSig
      }));
      this.setMessageStatus(messageId, MessageStatus.SIGNED);
      return rawSig;
    } catch (error) {
      log.error(error);
      msgObject.error = (error === null || error === void 0 ? void 0 : error.message) || (error === null || error === void 0 ? void 0 : error.toString());
      this.setMessageStatus(messageId, MessageStatus.FAILED);
    }
  }
  async addNewUnapprovedMessage(messageParams, req) {
    await this.addUnapprovedMessage(messageParams, req);
    return this.waitForFinishStatus(messageParams, this.name);
  }
  async addUnapprovedMessage(messageParams, req) {
    validateSignMessageData(messageParams);
    if (req) {
      messageParams.origin = req.origin;
    }
    messageParams.data = normalizeMessageData(messageParams.data);
    const messageId = messageParams.id || randomId();
    const messageData = {
      id: messageId,
      messageParams,
      status: MessageStatus.UNAPPROVED,
      time: Date.now(),
      type: METHOD_TYPES.ETH_SIGN
    };
    await this.addMessage(messageData);
    this.emit(MESSAGE_EVENTS.UNAPPROVED_MESSAGE, {
      messageData,
      req
    });
    return messageId;
  }
  prepMessageForSigning(messageParams) {
    return Promise.resolve(messageParams);
  }
}

export { MessageController };
