// @toruslabs/base-controllers/dist/lib.esm/Popup/PopupHandler.js
import { BaseController } from '../BaseController.js';
import { FEATURES_DEFAULT_POPUP_WINDOW } from '../enums.js';
import { StreamWindow } from './StreamWindow.js';
import { getPopupFeatures } from '../utils/utils.js';

/*
Scenarios:
1. Open a normal popup window and no communication with it - Use PopupHandler
2. Open a popup window and communicate with it - Use PopupWithBcHandler (can initiate communication by waiting for window to open or not)

3. If window is already opened, pass in windowId to the popup handler. 
   This will establish communication with the popup window and sends it a new url to redirect to


If you're trying to open a window and it gets blocked (happens if you're in iframe or delay b/w click and opening window),
  StreamWindow is invoked and it writes in a channel to display a message to the user

Once user clicks on that modal/dialog, we pre-open the window and pass in the windowId (goes to 3)
*/

/**
 * Handles popup window management.
 * For broadcast channel communication, use url with `instanceId` coded into state parameter.
 * This state parameter will be passed across redirects according to OAuth spec.
 */
class PopupHandler extends BaseController {
  constructor({
    config,
    state
  }) {
    super({
      config,
      state
    });
    // this.id = randomId()
    // Add in dapp storage key to all popups as a hash parameter
    this.defaultConfig = {
      features: getPopupFeatures(FEATURES_DEFAULT_POPUP_WINDOW),
      target: "_blank",
      communicationEngine: null,
      communicationWindowManager: null,
      timeout: 30000,
      instanceId: "",
      handleWindowBlockAlert: null,
      isWindowAlreadyOpen: false
    };
    this.defaultState = {
      windowTimer: null,
      window: null,
      iClosedWindow: false,
      windowId: "",
      url: state.url
    };
    this.initialize();
    this._setupTimer();
  }
  async open() {
    // if window is already open
    const {
      target,
      features,
      communicationEngine,
      communicationWindowManager
    } = this.config;
    const {
      windowId,
      url
    } = this.state;
    // No window has been pre-opened
    if (!this.config.isWindowAlreadyOpen) {
      // try to open a window first
      let localWindow = window.open(url.href, target, features);
      if (!localWindow) {
        // if it's blocked, open StreamWindow
        const streamWindow = new StreamWindow({
          config: {
            communicationEngine,
            communicationWindowManager,
            instanceId: this.config.instanceId,
            handleWindowBlockAlert: this.config.handleWindowBlockAlert,
            isWindowAlreadyOpen: this.config.isWindowAlreadyOpen
          },
          state: {
            url,
            windowId
          }
        });
        streamWindow.open();
        localWindow = streamWindow;
      }
      this.update({
        window: localWindow
      });
      return;
    }
    // A window has been pre-opened with a query parameter `windowId`
    const localWindow = new StreamWindow({
      config: {
        communicationEngine,
        communicationWindowManager,
        instanceId: this.config.instanceId,
        handleWindowBlockAlert: this.config.handleWindowBlockAlert,
        isWindowAlreadyOpen: this.config.isWindowAlreadyOpen
      },
      state: {
        url,
        windowId
      }
    });
    this.update({
      window: localWindow
    });
    await localWindow.open();
  }
  close() {
    this.update({
      iClosedWindow: true
    });
    const {
      window
    } = this.state;
    if (window) window.close();
  }
  _setupTimer() {
    const timer = window.setInterval(() => {
      const {
        window,
        windowTimer,
        iClosedWindow
      } = this.state;
      if (window && window.closed) {
        if (windowTimer) clearInterval(windowTimer);
        setTimeout(() => {
          if (!iClosedWindow) {
            this.emit("close");
          }
          this.update({
            iClosedWindow: false,
            window: null
          });
        }, this.config.timeout);
      }
      if (window === null && windowTimer) clearInterval(windowTimer);
    }, 500);
    this.update({
      windowTimer: timer
    });
  }
}

export { PopupHandler };
