// @web3auth/no-modal/dist/lib.esm/providers/xrpl-provider/providers/privateKeyProviders/xrplWalletUtils.js
import { rpcErrors, providerErrors } from '@web3auth/auth';
import { sign, generateSeed } from 'ripple-keypairs';
import { Client, Wallet, deriveAddress } from 'xrpl';
import ECDSA from 'xrpl/dist/npm/ECDSA';

const deriveKeypair = web3authKey => {
  const seed = generateSeed({
    entropy: Buffer.from(web3authKey.padStart(64, "0"), "hex"),
    algorithm: "ecdsa-secp256k1"
  });
  const wallet = Wallet.fromSecret(seed, {
    algorithm: ECDSA.secp256k1
  }); // web3auth network currently only supports the secp256k1 key
  return {
    privateKey: wallet.privateKey,
    publicKey: wallet.publicKey
  };
};
async function getProviderHandlers({
  privKey: web3authKey,
  chainConfig,
  keyExportEnabled
}) {
  const client = new Client(chainConfig.wsTarget);
  await client.connect();
  return {
    getAccounts: async _ => {
      const {
        publicKey
      } = deriveKeypair(web3authKey);
      const accAddress = deriveAddress(publicKey);
      return [accAddress];
    },
    getKeyPair: async _ => {
      if (!keyExportEnabled) throw providerErrors.custom({
        message: "Private key export is disabled",
        code: 4902
      });
      return deriveKeypair(web3authKey);
    },
    getPublicKey: async _ => {
      const keyPair = deriveKeypair(web3authKey);
      return keyPair.publicKey;
    },
    signTransaction: async req => {
      const {
        transaction,
        multisign
      } = req.params || {};
      if (!transaction) throw rpcErrors.invalidParams("Invalid params, req.params.transaction is required");
      const {
        publicKey,
        privateKey
      } = deriveKeypair(web3authKey);
      const wallet = new Wallet(publicKey, privateKey);
      return wallet.sign(transaction, multisign);
    },
    submitTransaction: async req => {
      const {
        transaction
      } = req.params || {};
      if (!transaction) throw rpcErrors.invalidParams("Invalid params, req.params.transaction is required");
      const {
        publicKey,
        privateKey
      } = deriveKeypair(web3authKey);
      const wallet = new Wallet(publicKey, privateKey);
      const res = await client.submit(transaction, {
        wallet
      });
      return res;
    },
    signMessage: async req => {
      const {
        message
      } = req.params || {};
      if (!message) throw rpcErrors.invalidParams("Invalid params, req.params.message is required");
      const keyPair = deriveKeypair(web3authKey);
      const signature = sign(message, keyPair.privateKey);
      return {
        signature
      };
    }
  };
}

export { getProviderHandlers };
