// @web3auth/modal/dist/lib.cjs/packages/modal/src/ui/components/ConnectWallet/ConnectWalletChainFilter/ConnectWalletChainFilter.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var noModal = require('@web3auth/no-modal');
var react = require('react');
var reactI18next = require('react-i18next');
var localeImport = require('../../../localeImport.js');
var utils = require('../../../utils.js');

function ConnectWalletChainFilter(props) {
  const {
    isDark,
    isLoading,
    selectedChain,
    setSelectedChain,
    chainNamespace
  } = props;
  const [t] = reactI18next.useTranslation(undefined, {
    i18n: localeImport
  });
  const chains = react.useMemo(() => {
    const chains = [{
      id: "all",
      name: "modal.allChains",
      icon: ""
    }];
    for (const chain of chainNamespace) {
      if (chain === noModal.CHAIN_NAMESPACES.EIP155 || chain === noModal.CHAIN_NAMESPACES.SOLANA) {
        chains.push({
          id: chain,
          name: chain === noModal.CHAIN_NAMESPACES.EIP155 ? "EVM" : chain,
          icon: chain === noModal.CHAIN_NAMESPACES.EIP155 ? "ethereum" : chain
        });
      }
    }
    return chains;
  }, [chainNamespace]);
  if (isLoading) {
    return jsxRuntime.jsx("div", {
      className: "w3a--flex w3a--items-center w3a--justify-between w3a--gap-x-2",
      children: Array.from({
        length: chains.length
      }).map((_, index) => jsxRuntime.jsx("div", {
        className: "w3a--h-12 w3a--w-[100px] w3a--animate-pulse w3a--rounded-2xl w3a--bg-app-gray-200 dark:w3a--bg-app-gray-700"
      }, `chain-loader-${index}`))
    });
  }
  return jsxRuntime.jsx("div", {
    className: "w3a--flex w3a--items-center w3a--justify-items-start w3a--gap-x-4",
    children: chains.map(chain => jsxRuntime.jsxs("button", {
      type: "button",
      className: utils.cn("w3a--flex w3a--w-[104px] w3a--items-center w3a--justify-center w3a--gap-x-1 w3a--text-xs w3a--font-medium w3a--px-4 w3a--py-3 w3a--text-app-gray-500 dark:w3a--text-app-gray-300 hover:w3a--bg-app-gray-200 dark:hover:w3a--bg-app-gray-700 w3a--h-12 w3a--rounded-2xl w3a--border w3a--border-transparent", {
        "w3a--bg-app-gray-100 dark:w3a--bg-app-gray-800 w3a--border-app-gray-200 dark:w3a--border-app-gray-700 w3a--text-app-gray-900 dark:w3a--text-app-white hover:!w3a--bg-app-gray-100 dark:hover:!w3a--bg-app-gray-800": selectedChain === chain.id
      }),
      onClick: () => setSelectedChain(chain.id),
      children: [chain.icon && jsxRuntime.jsx("img", {
        src: utils.getIcons(isDark ? `${chain.icon}-dark` : `${chain.icon}-light`),
        alt: chain.name,
        className: "w3a--size-5 w3a--object-contain"
      }), jsxRuntime.jsx("span", {
        className: "first-letter:w3a--capitalize",
        children: t(chain.name)
      })]
    }, chain.id))
  });
}

module.exports = ConnectWalletChainFilter;
