// @web3auth/auth/dist/lib.cjs/starkey/utils.js
'use strict';

var util = require('@ethereumjs/util');
var starkwareCrypto = require('@toruslabs/starkware-crypto');
var bip39 = require('bip39');
var encUtils = require('enc-utils');

const STARKNET_NETWORKS = {
  mainnet: "mainnet",
  testnet: "testnet"
};
const STARKNET_NETWORK_ID_MAP = {
  [STARKNET_NETWORKS.mainnet]: 0,
  [STARKNET_NETWORKS.testnet]: 100001
};
function isHexPrefixed(str) {
  return (str || "").substring(0, 2) === "0x";
}

/*
 Returns an integer from a given section of bits out of a hex string.
 hex is the target hex string to slice.
 start represents the index of the first bit to cut from the hex string (binary) in LSB order.
 end represents the index of the last bit to cut from the hex string.
*/
function getIntFromBits(hex, start, end) {
  const bin = encUtils.hexToBinary(hex);
  const bits = bin.slice(start, end);
  const int = encUtils.binaryToNumber(bits);
  return int;
}

/**
 * param- privKey secp256k1 private key in hex format
 * param- accountIndex accountIndex represents an index of the possible associated wallets derived from the seed.
 * param- starknetType corresponding startnet network (refer to STARKNET_NETWORKS type for possible values)
 * returns Calculates the stark key pair based on the STARTNET_NETWORK_TYPE and a given index.
 */
function getStarkHDAccount(privKey, accountIndex, starknetType) {
  if (!STARKNET_NETWORK_ID_MAP[starknetType]) {
    throw new Error(`Invalid starknet network specified:- ${starknetType}`);
  }
  const privKeyBuffer = Buffer.from(privKey, "hex");
  if (privKeyBuffer.length !== 32) {
    throw new Error("Invalid privKey size");
  }
  const ethAddress = Buffer.from(util.privateToAddress(privKeyBuffer)).toString("hex");
  const sanitizedEthAddr = isHexPrefixed(ethAddress) ? ethAddress : `0x${ethAddress}`;
  const mnemonic = bip39.entropyToMnemonic(privKey);
  const application = STARKNET_NETWORK_ID_MAP[starknetType];

  // Draws the 31 LSBs of the eth address.
  const ethAddressInt1 = getIntFromBits(sanitizedEthAddr, -31);
  // Draws the following 31 LSBs of the eth address.
  const ethAddressInt2 = getIntFromBits(sanitizedEthAddr, -62, -31);
  const accountPath = `m/2645'/1195502025'/${application}'/${ethAddressInt1}'/${ethAddressInt2}'/${accountIndex}`;
  const keyPair = starkwareCrypto.getKeyPairFromPath(mnemonic, accountPath);
  return keyPair;
}

exports.STARKNET_NETWORKS = STARKNET_NETWORKS;
exports.getStarkHDAccount = getStarkHDAccount;
