// @web3auth/modal/dist/lib.esm/packages/modal/src/ui/handlers/SmsPasswordlessHandler.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { PasswordlessHandler } from './AbstractHandler.js';

class SmsPasswordlessHandler extends PasswordlessHandler {
  constructor(params) {
    super(params);
    if (this.sessionStorageAvailable) {
      var _window$sessionStorag;
      this.trackingId = (_window$sessionStorag = window.sessionStorage.getItem("trackingId")) !== null && _window$sessionStorag !== void 0 ? _window$sessionStorag : undefined;
    }
  }
  async sendVerificationCode({
    captchaToken
  }) {
    const {
      loginHint,
      network,
      web3authClientId
    } = this.passwordlessParams;
    const finalParams = _objectSpread({
      client_id: web3authClientId,
      web3auth_client_id: web3authClientId,
      connection: this.connection,
      login_hint: loginHint,
      tracking_id: this.trackingId,
      whitelabel: this.whiteLabelParams,
      version: this.version,
      network
    }, captchaToken && {
      captcha_token: captchaToken
    });
    return super.start(finalParams);
  }
  async verifyCode(code) {
    const {
      loginHint,
      network,
      web3authClientId
    } = this.passwordlessParams;
    const params = {
      client_id: web3authClientId,
      login_hint: loginHint,
      code,
      connection: this.connection,
      tracking_id: this.trackingId,
      version: this.version,
      network
    };
    return super.verify(params);
  }
}

export { SmsPasswordlessHandler as default };
