// @toruslabs/base-controllers/dist/lib.cjs/utils/PollingManager.js
'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var log = require('loglevel');
var utils = require('./utils.js');

class PollingManager {
  constructor(idleTimeTracker, pollingInterval) {
    _defineProperty(this, "idleTimeTracker", void 0);
    _defineProperty(this, "pollingInterval", void 0);
    _defineProperty(this, "isPolling", false);
    this.pollingInterval = pollingInterval;
    this.idleTimeTracker = idleTimeTracker;
  }
  start(action) {
    this.idleTimeTracker.on("active", this.poll.bind(this, action));
    this.idleTimeTracker.on("idle", this.stop.bind(this));
    this.poll(action);
  }
  stop() {
    this.isPolling = false;
  }
  async poll(action) {
    if (this.isPolling) return;
    this.isPolling = true;
    while (this.isPolling) {
      if (this.idleTimeTracker.checkIfIdle()) {
        this.isPolling = false;
        return;
      }
      try {
        await action();
      } catch (error) {
        log.error(error);
      }
      await utils.timeout(this.pollingInterval);
    }
  }
}

exports.PollingManager = PollingManager;
