// @web3auth/modal/dist/lib.cjs/packages/modal/src/ui/components/Footer/Footer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var reactI18next = require('react-i18next');
var localeImport = require('../../localeImport.js');

/**
 * Footer component
 * @returns Footer component
 */
function Footer({
  privacyPolicy,
  termsOfService
}) {
  const [t] = reactI18next.useTranslation(undefined, {
    i18n: localeImport
  });
  return jsxRuntime.jsxs("div", {
    className: "w3a--mx-auto w3a--mt-auto w3a--flex w3a--flex-col w3a--items-center w3a--justify-center w3a--gap-y-4 w3a--pt-5",
    children: [(privacyPolicy || termsOfService) && jsxRuntime.jsxs("p", {
      className: "w3a--mx-auto w3a--w-4/5 w3a--text-center w3a--text-xs w3a--text-app-gray-500 dark:w3a--text-app-gray-400",
      children: [t("modal.footer.by-signing-in"), " ", termsOfService && jsxRuntime.jsxs("a", {
        href: termsOfService,
        className: "w3a--text-app-primary-600 dark:w3a--text-app-primary-500",
        children: [t("modal.footer.terms-of-service"), " "]
      }), privacyPolicy && jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [t("modal.footer.and"), " ", jsxRuntime.jsx("a", {
          href: privacyPolicy,
          className: "w3a--text-app-primary-600 dark:w3a--text-app-primary-500",
          children: t("modal.footer.privacy-policy")
        })]
      })]
    }), jsxRuntime.jsxs("div", {
      className: "w3a--flex w3a--items-center w3a--justify-center w3a--gap-2",
      children: [jsxRuntime.jsx("img", {
        height: "36",
        src: "https://images.web3auth.io/metamask-footer-logo-light.svg",
        alt: "Web3Auth Logo Light",
        className: "w3a--block w3a--h-9 dark:w3a--hidden"
      }), jsxRuntime.jsx("img", {
        height: "36",
        src: "https://images.web3auth.io/metamask-footer-logo-dark.svg",
        alt: "Web3Auth Logo Dark",
        className: "w3a--hidden w3a--h-9 dark:w3a--block"
      })]
    })]
  });
}

module.exports = Footer;
