// @toruslabs/ethereum-controllers/dist/lib.cjs/AccountAbstraction/smartAccounts/SafeSmartAccount.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var accounts = require('permissionless/accounts');
var accountAbstraction = require('viem/account-abstraction');
var constants = require('../../utils/constants.js');

class SafeSmartAccount {
  constructor(options) {
    _defineProperty(this, "name", constants.SMART_ACCOUNT.SAFE);
    _defineProperty(this, "options", void 0);
    this.options = options;
  }
  async getSmartAccount(params) {
    var _this$options, _this$options2, _this$options3;
    return accounts.toSafeSmartAccount(_objectSpread(_objectSpread(_objectSpread({}, this.options || {}), {}, {
      entryPoint: {
        address: ((_this$options = this.options) === null || _this$options === void 0 || (_this$options = _this$options.entryPoint) === null || _this$options === void 0 ? void 0 : _this$options.address) || accountAbstraction.entryPoint07Address,
        version: ((_this$options2 = this.options) === null || _this$options2 === void 0 || (_this$options2 = _this$options2.entryPoint) === null || _this$options2 === void 0 ? void 0 : _this$options2.version) || "0.7"
      },
      version: ((_this$options3 = this.options) === null || _this$options3 === void 0 ? void 0 : _this$options3.version) || "1.4.1"
    }, params), {}, {
      owners: [params.walletClient],
      client: params.client
    }));
  }
}

exports.SafeSmartAccount = SafeSmartAccount;
