// @web3auth/base/dist/lib.cjs/wallet/index.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');

const MULTI_CHAIN_ADAPTERS = {
  AUTH: "auth",
  WALLET_CONNECT_V2: "wallet-connect-v2",
  SFA: "sfa"
};
const SOLANA_ADAPTERS = _objectSpread({
  TORUS_SOLANA: "torus-solana"
}, MULTI_CHAIN_ADAPTERS);
const EVM_ADAPTERS = _objectSpread({
  TORUS_EVM: "torus-evm",
  COINBASE: "coinbase"
}, MULTI_CHAIN_ADAPTERS);
const WALLET_ADAPTERS = _objectSpread(_objectSpread({}, EVM_ADAPTERS), SOLANA_ADAPTERS);
const ADAPTER_NAMES = {
  [MULTI_CHAIN_ADAPTERS.AUTH]: "Auth",
  [MULTI_CHAIN_ADAPTERS.WALLET_CONNECT_V2]: "Wallet Connect v2",
  [MULTI_CHAIN_ADAPTERS.SFA]: "SFA",
  [SOLANA_ADAPTERS.TORUS_SOLANA]: "Torus",
  [EVM_ADAPTERS.TORUS_EVM]: "Torus",
  [EVM_ADAPTERS.COINBASE]: "Coinbase Smart Wallet"
};

exports.ADAPTER_NAMES = ADAPTER_NAMES;
exports.EVM_ADAPTERS = EVM_ADAPTERS;
exports.MULTI_CHAIN_ADAPTERS = MULTI_CHAIN_ADAPTERS;
exports.SOLANA_ADAPTERS = SOLANA_ADAPTERS;
exports.WALLET_ADAPTERS = WALLET_ADAPTERS;
