// @toruslabs/ethereum-controllers/dist/lib.cjs/Network/cacheIdentifier.js
'use strict';

var stringify = require('fast-safe-stringify');

exports.CacheStrategy = void 0;
(function (CacheStrategy) {
  /**
   * Cache per-block.
   */
  CacheStrategy["Block"] = "block";
  /**
   * Cache until a chain reorganization occurs.
   */
  CacheStrategy["Fork"] = "fork";
  /**
   * Never cache.
   */
  CacheStrategy["Never"] = "never";
  /**
   * Permanently cache.
   */
  CacheStrategy["Permanent"] = "perma";
})(exports.CacheStrategy || (exports.CacheStrategy = {}));
function blockTagParamIndex(method) {
  switch (method) {
    // blockTag is at index 2
    case "eth_getStorageAt":
      return 2;
    // blockTag is at index 1
    case "eth_getBalance":
    case "eth_getCode":
    case "eth_getTransactionCount":
    case "eth_call":
      return 1;
    // blockTag is at index 0
    case "eth_getBlockByNumber":
      return 0;
    // there is no blockTag
    default:
      return undefined;
  }
}
function cacheTypeForMethod(method) {
  switch (method) {
    // cache permanently
    case "web3_clientVersion":
    case "web3_sha3":
    case "eth_protocolVersion":
    case "eth_getBlockTransactionCountByHash":
    case "eth_getUncleCountByBlockHash":
    case "eth_getCode":
    case "eth_getBlockByHash":
    case "eth_getTransactionByHash":
    case "eth_getTransactionByBlockHashAndIndex":
    case "eth_getTransactionReceipt":
    case "eth_getUncleByBlockHashAndIndex":
    case "eth_getCompilers":
    case "eth_compileLLL":
    case "eth_compileSolidity":
    case "eth_compileSerpent":
    case "shh_version":
    case "test_permaCache":
      return exports.CacheStrategy.Permanent;
    // cache until fork
    case "eth_getBlockByNumber":
    case "eth_getBlockTransactionCountByNumber":
    case "eth_getUncleCountByBlockNumber":
    case "eth_getTransactionByBlockNumberAndIndex":
    case "eth_getUncleByBlockNumberAndIndex":
    case "test_forkCache":
      return exports.CacheStrategy.Fork;
    // cache for block
    case "eth_gasPrice":
    case "eth_blockNumber":
    case "eth_getBalance":
    case "eth_getStorageAt":
    case "eth_getTransactionCount":
    case "eth_call":
    case "eth_estimateGas":
    case "eth_getFilterLogs":
    case "eth_getLogs":
    case "test_blockCache":
      return exports.CacheStrategy.Block;
    // never cache
    default:
      return exports.CacheStrategy.Never;
  }
}
function canCache(method) {
  return cacheTypeForMethod(method) !== exports.CacheStrategy.Never;
}
function paramsWithoutBlockTag(request) {
  if (!request.params) {
    return [];
  }
  const index = blockTagParamIndex(request.method);
  // Block tag param not passed.
  if (index === undefined || !Array.isArray(request.params) || index >= request.params.length) {
    return request.params;
  }
  // eth_getBlockByNumber has the block tag first, then the optional includeTx? param
  if (request.method === "eth_getBlockByNumber") {
    return request.params.slice(1);
  }
  return request.params.slice(0, index);
}
function cacheIdentifierForRequest(request, skipBlockRef) {
  var _request$params;
  const simpleParams = skipBlockRef ? paramsWithoutBlockTag(request) : (_request$params = request.params) !== null && _request$params !== void 0 ? _request$params : [];
  if (canCache(request.method)) {
    return `${request.method}:${stringify(simpleParams)}`;
  }
  return null;
}

exports.blockTagParamIndex = blockTagParamIndex;
exports.cacheIdentifierForRequest = cacheIdentifierForRequest;
exports.cacheTypeForMethod = cacheTypeForMethod;
