// @web3auth/no-modal/dist/lib.cjs/providers/account-abstraction-provider/index.js
'use strict';

var AccountAbstractionProvider = require('./providers/AccountAbstractionProvider.js');
var ethereumControllers = require('@toruslabs/ethereum-controllers');



exports.accountAbstractionProvider = AccountAbstractionProvider.accountAbstractionProvider;
exports.toEoaProvider = AccountAbstractionProvider.toEoaProvider;
Object.defineProperty(exports, "BiconomySmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.BiconomySmartAccount; }
});
Object.defineProperty(exports, "KernelSmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.KernelSmartAccount; }
});
Object.defineProperty(exports, "MetamaskSmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.MetamaskSmartAccount; }
});
Object.defineProperty(exports, "NexusSmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.NexusSmartAccount; }
});
Object.defineProperty(exports, "SMART_ACCOUNT", {
	enumerable: true,
	get: function () { return ethereumControllers.SMART_ACCOUNT; }
});
Object.defineProperty(exports, "SafeSmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.SafeSmartAccount; }
});
Object.defineProperty(exports, "TrustSmartAccount", {
	enumerable: true,
	get: function () { return ethereumControllers.TrustSmartAccount; }
});
