// @toruslabs/base-controllers/dist/lib.cjs/index.js
'use strict';

var BaseController = require('./BaseController.js');
var BaseBlockTracker = require('./Block/BaseBlockTracker.js');
var createEventEmitterProxy = require('./createEventEmitterProxy.js');
var createSwappableProxy = require('./createSwappableProxy.js');
var BaseCurrencyController = require('./Currency/BaseCurrencyController.js');
var BaseEmbedController = require('./Embed/BaseEmbedController.js');
var CommunicationMethodMiddleware = require('./Embed/CommunicationMethodMiddleware.js');
var CommunicationWindowManager = require('./Embed/CommunicationWindowManager.js');
var IEmbedController = require('./Embed/IEmbedController.js');
var enums = require('./enums.js');
var interfaces = require('./interfaces.js');
var BaseKeyringController = require('./Keyring/BaseKeyringController.js');
var AbstractMessagecontroller = require('./Message/AbstractMessagecontroller.js');
var constants = require('./Message/constants.js');
var createFetchMiddleware = require('./Network/createFetchMiddleware.js');
var createInflightCacheMiddleware = require('./Network/createInflightCacheMiddleware.js');
var createLoggerMiddleware = require('./Network/createLoggerMiddleware.js');
var createOriginMiddleware = require('./Network/createOriginMiddleware.js');
var INetworkController = require('./Network/INetworkController.js');
var BasePreferencesController = require('./Preferences/BasePreferencesController.js');
var IPreferencesController = require('./Preferences/IPreferencesController.js');
var BaseTransactionStateController = require('./Transaction/BaseTransactionStateController.js');
var ITransactionController = require('./Transaction/ITransactionController.js');
var WSApiClient = require('./WSApiClient.js');
var BroadcastChannelHandler = require('./Popup/BroadcastChannelHandler.js');
var PopupHandler = require('./Popup/PopupHandler.js');
var PopupStoreChannel = require('./Popup/PopupStoreChannel.js');
var PopupWithBcHandler = require('./Popup/PopupWithBcHandler.js');
var RedirectHandler = require('./Popup/RedirectHandler.js');
var StreamWindow = require('./Popup/StreamWindow.js');
var helpers = require('./utils/helpers.js');
var lodashUtils = require('./utils/lodashUtils.js');
var signingUtils = require('./utils/signingUtils.js');
var txUtils = require('./utils/txUtils.js');
var utils = require('./utils/utils.js');



exports.BaseController = BaseController.BaseController;
exports.BaseBlockTracker = BaseBlockTracker.BaseBlockTracker;
exports.createEventEmitterProxy = createEventEmitterProxy.createEventEmitterProxy;
exports.createSwappableProxy = createSwappableProxy.createSwappableProxy;
exports.BaseCurrencyController = BaseCurrencyController.BaseCurrencyController;
exports.BaseEmbedController = BaseEmbedController.BaseEmbedController;
exports.createChangeProviderMiddlewareMiddleware = CommunicationMethodMiddleware.createChangeProviderMiddlewareMiddleware;
exports.createCommunicationMiddleware = CommunicationMethodMiddleware.createCommunicationMiddleware;
exports.createGenericJRPCMiddleware = CommunicationMethodMiddleware.createGenericJRPCMiddleware;
exports.createTopupMiddleware = CommunicationMethodMiddleware.createTopupMiddleware;
exports.CommunicationWindowManager = CommunicationWindowManager.CommunicationWindowManager;
exports.BUTTON_POSITION = IEmbedController.BUTTON_POSITION;
exports.CONFIRMATION_STRATEGY = IEmbedController.CONFIRMATION_STRATEGY;
exports.ACTIVITY_ACTION_ACCEPT_NFT_OFFER = enums.ACTIVITY_ACTION_ACCEPT_NFT_OFFER;
exports.ACTIVITY_ACTION_ALL = enums.ACTIVITY_ACTION_ALL;
exports.ACTIVITY_ACTION_BURN = enums.ACTIVITY_ACTION_BURN;
exports.ACTIVITY_ACTION_CANCEL_NFT_OFFER = enums.ACTIVITY_ACTION_CANCEL_NFT_OFFER;
exports.ACTIVITY_ACTION_CREATE_NFT_OFFER = enums.ACTIVITY_ACTION_CREATE_NFT_OFFER;
exports.ACTIVITY_ACTION_CREATE_TRUSTLINE = enums.ACTIVITY_ACTION_CREATE_TRUSTLINE;
exports.ACTIVITY_ACTION_RECEIVE = enums.ACTIVITY_ACTION_RECEIVE;
exports.ACTIVITY_ACTION_REMOVE_TRUSTLINE = enums.ACTIVITY_ACTION_REMOVE_TRUSTLINE;
exports.ACTIVITY_ACTION_SEND = enums.ACTIVITY_ACTION_SEND;
exports.ACTIVITY_ACTION_TOPUP = enums.ACTIVITY_ACTION_TOPUP;
exports.ACTIVITY_PERIOD_ALL = enums.ACTIVITY_PERIOD_ALL;
exports.ACTIVITY_PERIOD_MONTH_ONE = enums.ACTIVITY_PERIOD_MONTH_ONE;
exports.ACTIVITY_PERIOD_MONTH_SIX = enums.ACTIVITY_PERIOD_MONTH_SIX;
exports.ACTIVITY_PERIOD_WEEK_ONE = enums.ACTIVITY_PERIOD_WEEK_ONE;
exports.ACTIVITY_STATUS_CANCELLED = enums.ACTIVITY_STATUS_CANCELLED;
exports.ACTIVITY_STATUS_CANCELLING = enums.ACTIVITY_STATUS_CANCELLING;
exports.ACTIVITY_STATUS_PENDING = enums.ACTIVITY_STATUS_PENDING;
exports.ACTIVITY_STATUS_SUCCESSFUL = enums.ACTIVITY_STATUS_SUCCESSFUL;
exports.ACTIVITY_STATUS_UNSUCCESSFUL = enums.ACTIVITY_STATUS_UNSUCCESSFUL;
exports.BROADCAST_CHANNELS = enums.BROADCAST_CHANNELS;
exports.BROADCAST_CHANNELS_MSGS = enums.BROADCAST_CHANNELS_MSGS;
exports.COMMUNICATION_JRPC_METHODS = enums.COMMUNICATION_JRPC_METHODS;
exports.COMMUNICATION_NOTIFICATIONS = enums.COMMUNICATION_NOTIFICATIONS;
exports.ControllerEvents = enums.ControllerEvents;
exports.FEATURES_CONFIRM_WINDOW = enums.FEATURES_CONFIRM_WINDOW;
exports.FEATURES_DEFAULT_POPUP_WINDOW = enums.FEATURES_DEFAULT_POPUP_WINDOW;
exports.FEATURES_DEFAULT_WALLET_WINDOW = enums.FEATURES_DEFAULT_WALLET_WINDOW;
exports.FEATURES_PROVIDER_CHANGE_WINDOW = enums.FEATURES_PROVIDER_CHANGE_WINDOW;
exports.POPUP_LOADED = enums.POPUP_LOADED;
exports.POPUP_RESULT = enums.POPUP_RESULT;
exports.PROVIDER_JRPC_METHODS = enums.PROVIDER_JRPC_METHODS;
exports.PROVIDER_NOTIFICATIONS = enums.PROVIDER_NOTIFICATIONS;
exports.SETUP_COMPLETE = enums.SETUP_COMPLETE;
exports.LOGIN_PROVIDER = interfaces.LOGIN_PROVIDER;
exports.PAYMENT_PROVIDER = interfaces.PAYMENT_PROVIDER;
exports.BaseKeyringController = BaseKeyringController.BaseKeyringController;
exports.AbstractMessageController = AbstractMessagecontroller.AbstractMessageController;
exports.MESSAGE_EVENTS = constants.MESSAGE_EVENTS;
exports.MessageStatus = constants.MessageStatus;
exports.createFetchConfigFromReq = createFetchMiddleware.createFetchConfigFromReq;
exports.createFetchMiddleware = createFetchMiddleware.createFetchMiddleware;
exports.createInflightCacheMiddleware = createInflightCacheMiddleware.createInflightCacheMiddleware;
exports.createLoggerMiddleware = createLoggerMiddleware.createLoggerMiddleware;
exports.createOriginMiddleware = createOriginMiddleware.createOriginMiddleware;
exports.CHAIN_NAMESPACES = INetworkController.CHAIN_NAMESPACES;
exports.createRandomId = INetworkController.createRandomId;
exports.BasePreferencesController = BasePreferencesController.BasePreferencesController;
exports.DEFAULT_PREFERENCES = BasePreferencesController.DEFAULT_PREFERENCES;
exports.ACCOUNT_CATEGORY = IPreferencesController.ACCOUNT_CATEGORY;
exports.ACTIVITY_ACTION = IPreferencesController.ACTIVITY_ACTION;
exports.BaseTransactionStateManager = BaseTransactionStateController.BaseTransactionStateManager;
exports.TRANSACTION_TYPES = ITransactionController.TRANSACTION_TYPES;
exports.TX_EVENTS = ITransactionController.TX_EVENTS;
exports.TransactionStatus = ITransactionController.TransactionStatus;
exports.WSApiClient = WSApiClient.WSApiClient;
exports.BroadcastChannelHandler = BroadcastChannelHandler.BroadcastChannelHandler;
exports.PopupHandler = PopupHandler.PopupHandler;
exports.PopupStoreChannel = PopupStoreChannel.PopupStoreChannel;
exports.PopupWithBcHandler = PopupWithBcHandler.PopupWithBcHandler;
exports.RedirectHandler = RedirectHandler.RedirectHandler;
exports.StreamWindow = StreamWindow.StreamWindow;
exports.DEFAULT_IDLE_TIME = helpers.DEFAULT_IDLE_TIME;
exports.createIdleTracker = helpers.createIdleTracker;
exports.idleTimeTracker = helpers.idleTimeTracker;
exports.cloneDeep = lodashUtils.cloneDeep;
exports.omitBy = lodashUtils.omitBy;
exports.pickBy = lodashUtils.pickBy;
exports.authServer = signingUtils.authServer;
exports.signChallenge = signingUtils.signChallenge;
exports.verifySignedChallenge = signingUtils.verifySignedChallenge;
exports.getTxStatusText = txUtils.getTxStatusText;
exports.UserError = utils.UserError;
exports.addressSlicer = utils.addressSlicer;
exports.broadcastChannelOptions = utils.broadcastChannelOptions;
exports.concatSig = utils.concatSig;
exports.formatDate = utils.formatDate;
exports.formatSmallNumbers = utils.formatSmallNumbers;
exports.formatTime = utils.formatTime;
exports.getCustomDeviceInfo = utils.getCustomDeviceInfo;
exports.getHeaders = utils.getHeaders;
exports.getPopupFeatures = utils.getPopupFeatures;
exports.handleRedirectParameters = utils.handleRedirectParameters;
exports.hashMessage = utils.hashMessage;
exports.intToHex = utils.intToHex;
exports.isUnauthorizedError = utils.isUnauthorizedError;
exports.padWithZeroes = utils.padWithZeroes;
exports.randomId = utils.randomId;
exports.signMessage = utils.signMessage;
exports.significantDigits = utils.significantDigits;
exports.sleep = utils.sleep;
exports.timeout = utils.timeout;
exports.transactionMatchesNetwork = utils.transactionMatchesNetwork;
