// @toruslabs/base-controllers/dist/lib.cjs/Preferences/IPreferencesController.js
'use strict';

const ACTIVITY_ACTION = {
  ACTIVITY_ACTION_ALL: "walletActivity.allTransactions",
  ACTIVITY_ACTION_SEND: "walletActivity.send",
  ACTIVITY_ACTION_RECEIVE: "walletActivity.receive",
  ACTIVITY_ACTION_TOPUP: "walletActivity.topup"
};
const ACCOUNT_CATEGORY = {
  NORMAL: "normal",
  THRESHOLD: "threshold",
  IMPORTED: "imported",
  // we have private key here
  APP_SCOPED: "app_scoped",
  ACCOUNT_ABSTRACTION: "account_abstraction",
  EXTERNAL: "external",
  // like metamask, wallet connect
  MPC: "mpc",
  SFA: "sfa",
  APP_SCOPED_DERIVED: "app_scoped_derived"
};

/**
 * Preferences controller state
 */

/**
 * @param address - address of the user
 * @param jwtToken - if jwt token already exists, it will be used instead of fetching from backend
 * @param calledFromEmbed - if called from embed
 * @param userInfo - optional user info
 * @param rehydrate - Rehydrate the preferences from the local storage
 */

exports.ACCOUNT_CATEGORY = ACCOUNT_CATEGORY;
exports.ACTIVITY_ACTION = ACTIVITY_ACTION;
