// @web3auth/solana-provider/dist/lib.cjs/index.js
'use strict';

var solanaWallet = require('./solanaWallet.js');
var torusInjectedProvider = require('./providers/injectedProviders/torus/torusInjectedProvider.js');
var walletStandardProvider = require('./providers/injectedProviders/walletStandardProvider.js');
var solanaPrivateKeyProvider = require('./providers/privateKeyProvider/solanaPrivateKeyProvider.js');
var JrpcClient = require('./rpc/JrpcClient.js');
var solanaRpcMiddlewares = require('./rpc/solanaRpcMiddlewares.js');



exports.SolanaWallet = solanaWallet.SolanaWallet;
exports.TorusInjectedProvider = torusInjectedProvider.TorusInjectedProvider;
exports.WalletStandardProvider = walletStandardProvider.WalletStandardProvider;
exports.SolanaPrivateKeyProvider = solanaPrivateKeyProvider.SolanaPrivateKeyProvider;
exports.createChainIdMiddleware = JrpcClient.createChainIdMiddleware;
exports.createConfigMiddleware = JrpcClient.createConfigMiddleware;
exports.createJsonRpcClient = JrpcClient.createJsonRpcClient;
exports.createProviderConfigMiddleware = JrpcClient.createProviderConfigMiddleware;
exports.createAccountMiddleware = solanaRpcMiddlewares.createAccountMiddleware;
exports.createChainSwitchMiddleware = solanaRpcMiddlewares.createChainSwitchMiddleware;
exports.createGenericJRPCMiddleware = solanaRpcMiddlewares.createGenericJRPCMiddleware;
exports.createGetAccountsMiddleware = solanaRpcMiddlewares.createGetAccountsMiddleware;
exports.createGetPublicKeyMiddleware = solanaRpcMiddlewares.createGetPublicKeyMiddleware;
exports.createRequestAccountsMiddleware = solanaRpcMiddlewares.createRequestAccountsMiddleware;
exports.createSolanaMiddleware = solanaRpcMiddlewares.createSolanaMiddleware;
