// @web3auth/auth/dist/lib.cjs/index.js
'use strict';

var index = require('./starkey/index.js');
var logger = require('./utils/logger.js');
var auth = require('./core/auth.js');
var errors = require('./core/errors.js');
var utils = require('./core/utils.js');
var utils$1 = require('./ed25519/utils.js');
var basePostMessageStream = require('./jrpc/basePostMessageStream.js');
var postMessageStream = require('./jrpc/postMessageStream.js');
var safeEventEmitter = require('./jrpc/safeEventEmitter.js');
var serializableError = require('./jrpc/serializableError.js');
var substream = require('./jrpc/substream.js');
var errorClasses = require('./jrpc/errors/error-classes.js');
var errorConstants = require('./jrpc/errors/error-constants.js');
var errors$1 = require('./jrpc/errors/errors.js');
var utils$2 = require('./jrpc/errors/utils.js');
var jrpc = require('./jrpc/jrpc.js');
var jrpcEngine = require('./jrpc/jrpcEngine.js');
var mux = require('./jrpc/mux.js');
var utils$3 = require('./starkey/utils.js');
var keyDerivation = require('./subkey/keyDerivation.js');
var mimcsponge = require('./subkey/mimcsponge.js');
var browserStorage = require('./utils/browserStorage.js');
var constants = require('./utils/constants.js');
var interfaces = require('./utils/interfaces.js');
var utils$4 = require('./utils/utils.js');
var whitelabel = require('./utils/whitelabel.js');



exports.starkEc = index.starkEc;
exports.starkGrindKey = index.starkGrindKey;
exports.starkPedersen = index.starkPedersen;
exports.starkSign = index.starkSign;
exports.starkVerify = index.starkVerify;
exports.loglevel = logger.loglevel;
exports.Auth = auth.Auth;
exports.AuthError = errors.AuthError;
exports.InitializationError = errors.InitializationError;
exports.LoginError = errors.LoginError;
exports.constructURL = utils.constructURL;
exports.getHashQueryParams = utils.getHashQueryParams;
exports.getPopupFeatures = utils.getPopupFeatures;
exports.getTimeout = utils.getTimeout;
exports.isMobileOrTablet = utils.isMobileOrTablet;
exports.version = utils.version;
exports.getED25519Key = utils$1.getED25519Key;
exports.BasePostMessageStream = basePostMessageStream.BasePostMessageStream;
exports.PostMessageStream = postMessageStream.PostMessageStream;
exports.SafeEventEmitter = safeEventEmitter.SafeEventEmitter;
exports.SerializableError = serializableError.SerializableError;
exports.Substream = substream.Substream;
exports.EthereumProviderError = errorClasses.EthereumProviderError;
exports.JsonRpcError = errorClasses.JsonRpcError;
exports.errorCodes = errorConstants.errorCodes;
exports.errorValues = errorConstants.errorValues;
exports.providerErrors = errors$1.providerErrors;
exports.rpcErrors = errors$1.rpcErrors;
exports.JSON_RPC_SERVER_ERROR_MESSAGE = utils$2.JSON_RPC_SERVER_ERROR_MESSAGE;
exports.dataHasCause = utils$2.dataHasCause;
exports.getMessageFromCode = utils$2.getMessageFromCode;
exports.isObject = utils$2.isObject;
exports.isPlainObject = utils$2.isPlainObject;
exports.isValidCode = utils$2.isValidCode;
exports.isValidString = utils$2.isValidString;
exports.serializeCause = utils$2.serializeCause;
exports.serializeError = utils$2.serializeError;
exports.createAsyncMiddleware = jrpc.createAsyncMiddleware;
exports.createErrorMiddleware = jrpc.createErrorMiddleware;
exports.createIdRemapMiddleware = jrpc.createIdRemapMiddleware;
exports.createLoggerMiddleware = jrpc.createLoggerMiddleware;
exports.createScaffoldMiddleware = jrpc.createScaffoldMiddleware;
exports.createStreamMiddleware = jrpc.createStreamMiddleware;
exports.getRpcPromiseCallback = jrpc.getRpcPromiseCallback;
exports.JRPCEngine = jrpcEngine.JRPCEngine;
exports.createEngineStream = jrpcEngine.createEngineStream;
exports.mergeMiddleware = jrpcEngine.mergeMiddleware;
exports.providerAsMiddleware = jrpcEngine.providerAsMiddleware;
exports.providerFromEngine = jrpcEngine.providerFromEngine;
exports.providerFromMiddleware = jrpcEngine.providerFromMiddleware;
exports.IGNORE_SUBSTREAM = mux.IGNORE_SUBSTREAM;
exports.ObjectMultiplex = mux.ObjectMultiplex;
exports.setupMultiplex = mux.setupMultiplex;
exports.STARKNET_NETWORKS = utils$3.STARKNET_NETWORKS;
exports.getStarkHDAccount = utils$3.getStarkHDAccount;
exports.SECP256K1_CURVE_N = keyDerivation.SECP256K1_CURVE_N;
exports.subkey = keyDerivation.subkey;
exports.mimcGetConstants = mimcsponge.mimcGetConstants;
exports.mimcHash = mimcsponge.mimcHash;
exports.mimcMultiHash = mimcsponge.mimcMultiHash;
exports.mimgGetIV = mimcsponge.mimgGetIV;
exports.BrowserStorage = browserStorage.BrowserStorage;
exports.MemoryStore = browserStorage.MemoryStore;
exports.AUTH_ACTIONS = constants.AUTH_ACTIONS;
exports.BUILD_ENV = constants.BUILD_ENV;
exports.LOGIN_PROVIDER = constants.LOGIN_PROVIDER;
exports.MFA_LEVELS = constants.MFA_LEVELS;
exports.SUPPORTED_KEY_CURVES = constants.SUPPORTED_KEY_CURVES;
exports.UX_MODE = constants.UX_MODE;
exports.WEB3AUTH_LEGACY_NETWORK = constants.WEB3AUTH_LEGACY_NETWORK;
exports.WEB3AUTH_NETWORK = constants.WEB3AUTH_NETWORK;
exports.WEB3AUTH_SAPPHIRE_NETWORK = constants.WEB3AUTH_SAPPHIRE_NETWORK;
exports.storeKey = constants.storeKey;
exports.LANGUAGES = interfaces.LANGUAGES;
exports.LANGUAGE_MAP = interfaces.LANGUAGE_MAP;
exports.MFA_FACTOR = interfaces.MFA_FACTOR;
exports.THEME_MODES = interfaces.THEME_MODES;
exports.base64toJSON = utils$4.base64toJSON;
exports.base64url = utils$4.base64url;
exports.jsonToBase64 = utils$4.jsonToBase64;
exports.safeatob = utils$4.safeatob;
exports.safebtoa = utils$4.safebtoa;
exports.storageAvailable = utils$4.storageAvailable;
exports.applyWhiteLabelTheme = whitelabel.applyWhiteLabelTheme;
exports.generateWhiteLabelTheme = whitelabel.generateWhiteLabelTheme;
exports.getColorsList = whitelabel.getColorsList;
