// @web3auth/base/dist/lib.cjs/index.js
'use strict';

var config = require('./chain/config.js');
var IChainInterface = require('./chain/IChainInterface.js');
var index = require('./composables/index.js');
var index$1 = require('./errors/index.js');
var loglevel = require('./loglevel.js');
var IProvider = require('./provider/IProvider.js');
var utils$1 = require('./utils.js');
var index$2 = require('./wallet/index.js');
var auth = require('@web3auth/auth');
var IAdapter = require('./adapter/IAdapter.js');
var utils = require('./adapter/utils.js');
var errors = require('./plugin/errors.js');
var IPlugin = require('./plugin/IPlugin.js');
var baseControllers = require('@toruslabs/base-controllers');



exports.getChainConfig = config.getChainConfig;
exports.getEvmChainConfig = config.getEvmChainConfig;
exports.getSolanaChainConfig = config.getSolanaChainConfig;
exports.getXrplChainConfig = config.getXrplChainConfig;
exports.ADAPTER_NAMESPACES = IChainInterface.ADAPTER_NAMESPACES;
exports.CHAIN_NAMESPACES = IChainInterface.CHAIN_NAMESPACES;
exports.Web3AuthContextKey = index.Web3AuthContextKey;
exports.WalletInitializationError = index$1.WalletInitializationError;
exports.WalletLoginError = index$1.WalletLoginError;
exports.WalletOperationsError = index$1.WalletOperationsError;
exports.WalletProviderError = index$1.WalletProviderError;
exports.Web3AuthError = index$1.Web3AuthError;
exports.serializeError = index$1.serializeError;
exports.log = loglevel;
exports.PROVIDER_EVENTS = IProvider.PROVIDER_EVENTS;
exports.fetchProjectConfig = utils$1.fetchProjectConfig;
exports.fetchWalletRegistry = utils$1.fetchWalletRegistry;
exports.isHexStrict = utils$1.isHexStrict;
exports.normalizeWalletName = utils$1.normalizeWalletName;
exports.signerHost = utils$1.signerHost;
exports.storageAvailable = utils$1.storageAvailable;
exports.ADAPTER_NAMES = index$2.ADAPTER_NAMES;
exports.EVM_ADAPTERS = index$2.EVM_ADAPTERS;
exports.MULTI_CHAIN_ADAPTERS = index$2.MULTI_CHAIN_ADAPTERS;
exports.SOLANA_ADAPTERS = index$2.SOLANA_ADAPTERS;
exports.WALLET_ADAPTERS = index$2.WALLET_ADAPTERS;
Object.defineProperty(exports, "UX_MODE", {
	enumerable: true,
	get: function () { return auth.UX_MODE; }
});
Object.defineProperty(exports, "WEB3AUTH_NETWORK", {
	enumerable: true,
	get: function () { return auth.WEB3AUTH_NETWORK; }
});
exports.ADAPTER_CATEGORY = IAdapter.ADAPTER_CATEGORY;
exports.ADAPTER_EVENTS = IAdapter.ADAPTER_EVENTS;
exports.ADAPTER_STATUS = IAdapter.ADAPTER_STATUS;
exports.BaseAdapter = IAdapter.BaseAdapter;
exports.checkIfTokenIsExpired = utils.checkIfTokenIsExpired;
exports.clearToken = utils.clearToken;
exports.getSavedToken = utils.getSavedToken;
exports.saveToken = utils.saveToken;
exports.NFTCheckoutPluginError = errors.NFTCheckoutPluginError;
exports.WalletServicesPluginError = errors.WalletServicesPluginError;
exports.EVM_PLUGINS = IPlugin.EVM_PLUGINS;
exports.PLUGIN_EVENTS = IPlugin.PLUGIN_EVENTS;
exports.PLUGIN_NAMESPACES = IPlugin.PLUGIN_NAMESPACES;
exports.PLUGIN_STATUS = IPlugin.PLUGIN_STATUS;
exports.SOLANA_PLUGINS = IPlugin.SOLANA_PLUGINS;
exports.WALLET_PLUGINS = IPlugin.WALLET_PLUGINS;
Object.defineProperty(exports, "cloneDeep", {
	enumerable: true,
	get: function () { return baseControllers.cloneDeep; }
});
