// @segment/analytics-next/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveUserArguments = exports.resolvePageArguments = exports.resolveArguments = exports.resolveAliasArguments = exports.segmentio = exports.UniversalStorage = exports.getGlobalAnalytics = exports.AnalyticsBrowser = exports.Analytics = void 0;
var tslib_1 = require("tslib");
var analytics_1 = require("./core/analytics");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return analytics_1.Analytics; } });
var browser_1 = require("./browser");
Object.defineProperty(exports, "AnalyticsBrowser", { enumerable: true, get: function () { return browser_1.AnalyticsBrowser; } });
tslib_1.__exportStar(require("./node"), exports);
tslib_1.__exportStar(require("./core/context"), exports);
tslib_1.__exportStar(require("./core/events"), exports);
tslib_1.__exportStar(require("./core/plugin"), exports);
tslib_1.__exportStar(require("./core/user"), exports);
var global_analytics_helper_1 = require("./lib/global-analytics-helper");
Object.defineProperty(exports, "getGlobalAnalytics", { enumerable: true, get: function () { return global_analytics_helper_1.getGlobalAnalytics; } });
var storage_1 = require("./core/storage");
Object.defineProperty(exports, "UniversalStorage", { enumerable: true, get: function () { return storage_1.UniversalStorage; } });
var segmentio_1 = require("./plugins/segmentio");
Object.defineProperty(exports, "segmentio", { enumerable: true, get: function () { return segmentio_1.segmentio; } });
var arguments_resolver_1 = require("./core/arguments-resolver");
Object.defineProperty(exports, "resolveAliasArguments", { enumerable: true, get: function () { return arguments_resolver_1.resolveAliasArguments; } });
Object.defineProperty(exports, "resolveArguments", { enumerable: true, get: function () { return arguments_resolver_1.resolveArguments; } });
Object.defineProperty(exports, "resolvePageArguments", { enumerable: true, get: function () { return arguments_resolver_1.resolvePageArguments; } });
Object.defineProperty(exports, "resolveUserArguments", { enumerable: true, get: function () { return arguments_resolver_1.resolveUserArguments; } });
