// @web3auth/no-modal/dist/lib.cjs/base/constants.js
'use strict';

const SMART_ACCOUNT_WALLET_SCOPE = {
  EMBEDDED: "embedded",
  ALL: "all"
};
const MODAL_SIGN_IN_METHODS = {
  SOCIAL: "social",
  PASSWORDLESS: "passwordless",
  EXTERNAL_WALLETS: "externalWallets"
};
const WIDGET_TYPE = {
  MODAL: "modal",
  EMBED: "embed"
};
const WEB3AUTH_STATE_STORAGE_KEY = "Web3Auth-state";
const LOGIN_MODE = {
  MODAL: "modal",
  NO_MODAL: "no-modal"
};
const SOLANA_CAIP_CHAIN_MAP = {
  "0x65": "5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
  "0x66": "4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z",
  "0x67": "EtWTRABZaYq6iMfeYKouRu166VU2xqa1"
};
const WALLET_REGISTRY_URL = "https://assets.web3auth.io/v2/wallet-registry.json";

exports.LOGIN_MODE = LOGIN_MODE;
exports.MODAL_SIGN_IN_METHODS = MODAL_SIGN_IN_METHODS;
exports.SMART_ACCOUNT_WALLET_SCOPE = SMART_ACCOUNT_WALLET_SCOPE;
exports.SOLANA_CAIP_CHAIN_MAP = SOLANA_CAIP_CHAIN_MAP;
exports.WALLET_REGISTRY_URL = WALLET_REGISTRY_URL;
exports.WEB3AUTH_STATE_STORAGE_KEY = WEB3AUTH_STATE_STORAGE_KEY;
exports.WIDGET_TYPE = WIDGET_TYPE;
