// @web3auth/modal/dist/lib.esm/packages/modal/src/ui/components/Login/LoginPasswordLess/LoginPasswordLess.js
import { useRef, useState, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import i18nInstance from '../../../localeImport.js';
import { cn, getIcons } from '../../../utils.js';
import PulseLoader from '../../PulseLoader/PulseLoader.js';
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';

function LoginPasswordLess(props) {
  const {
    isModalVisible,
    isPasswordLessCtaClicked,
    setIsPasswordLessCtaClicked,
    title,
    placeholder,
    handleFormSubmit,
    errorMessage,
    isDark,
    isPasswordLessLoading,
    buttonRadius
  } = props;
  const [t] = useTranslation(undefined, {
    i18n: i18nInstance
  });
  const inputRef = useRef(null);
  const [inputValue, setInputValue] = useState("");
  const [isInputFocused, setIsInputFocused] = useState(false);
  const onInputChange = e => {
    setInputValue(e.target.value);
    setIsInputFocused(true);
  };
  const handleEnter = e => {
    const {
      key
    } = e;
    if (key === "Enter") {
      handleFormSubmit(inputValue);
      setIsInputFocused(false);
    }
  };
  const onFormSubmit = e => {
    e.preventDefault();
    e.stopPropagation();
    handleFormSubmit(inputValue);
    setIsInputFocused(false);
  };
  useEffect(() => {
    if (!isModalVisible) {
      setIsPasswordLessCtaClicked(false);
    }
  }, [isModalVisible, setIsPasswordLessCtaClicked]);
  useEffect(() => {
    if (isPasswordLessCtaClicked) {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 || _inputRef$current.focus();
    }
  }, [isPasswordLessCtaClicked]);
  if (!isPasswordLessCtaClicked) {
    return /*#__PURE__*/jsxs("button", {
      type: "button",
      className: cn("w3a--btn !w3a--justify-between w3a--relative w3a--group w3a--overflow-hidden", {
        "w3a--rounded-full": buttonRadius === "pill",
        "w3a--rounded-lg": buttonRadius === "rounded",
        "w3a--rounded-none": buttonRadius === "square"
      }),
      onClick: () => {
        setIsPasswordLessCtaClicked(true);
      },
      children: [/*#__PURE__*/jsx("p", {
        className: "w3a--text-app-gray-900 dark:w3a--text-app-white",
        children: t("modal.passwordless.title", {
          title
        })
      }), /*#__PURE__*/jsx("img", {
        id: "passwordless-arrow",
        className: "w3a--absolute w3a--right-4 w3a--top-1/2 -w3a--translate-x-6 -w3a--translate-y-1/2 w3a--opacity-0 w3a--transition-all w3a--duration-300 group-hover:w3a--translate-x-0 group-hover:w3a--opacity-100",
        src: getIcons(isDark ? "chevron-right-dark" : "chevron-right-light"),
        alt: "arrow"
      })]
    });
  }
  return /*#__PURE__*/jsxs(Fragment, {
    children: [/*#__PURE__*/jsxs("div", {
      className: cn("w3a--input", isInputFocused && "!w3a--border-app-primary-600", {
        "w3a--rounded-full": buttonRadius === "pill",
        "w3a--rounded-lg": buttonRadius === "rounded",
        "w3a--rounded-none": buttonRadius === "square"
      }),
      children: [/*#__PURE__*/jsx("input", {
        ref: inputRef,
        onInput: onInputChange,
        value: inputValue,
        placeholder: placeholder,
        onFocus: () => {
          setIsInputFocused(true);
        },
        onBlur: e => {
          e.target.placeholder = `${placeholder}`;
          setIsInputFocused(false);
        },
        type: "text",
        className: cn("w3a--w-full w3a--appearance-none w3a--bg-transparent w3a--text-app-gray-900 w3a--outline-none placeholder:w3a--text-xs placeholder:w3a--text-app-gray-400 focus:w3a--outline-none active:w3a--outline-none dark:w3a--text-app-white dark:placeholder:w3a--text-app-gray-500"),
        onKeyDown: e => handleEnter(e),
        disabled: isPasswordLessLoading
      }), isPasswordLessLoading && /*#__PURE__*/jsx(PulseLoader, {}), inputValue && !isPasswordLessLoading && /*#__PURE__*/jsx("button", {
        type: "button",
        className: "w3a--appearance-none",
        onClick: onFormSubmit,
        children: /*#__PURE__*/jsx("img", {
          src: getIcons(isDark ? "chevron-right-dark" : "chevron-right-light"),
          alt: "arrow"
        })
      })]
    }), errorMessage && !isInputFocused && isPasswordLessCtaClicked && /*#__PURE__*/jsx("p", {
      className: "w3a--w-full w3a--pl-6 w3a--text-start w3a--text-xs w3a--font-normal w3a--text-app-red-500 dark:w3a--text-app-red-400",
      children: errorMessage
    })]
  });
}

export { LoginPasswordLess as default };
