// @toruslabs/ethereum-controllers/dist/lib.cjs/Message/AddChainController.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var baseControllers = require('@toruslabs/base-controllers');
var log = require('loglevel');
var constants = require('../utils/constants.js');
var utils = require('./utils.js');

class AddChainController extends baseControllers.AbstractMessageController {
  constructor({
    config,
    state,
    addChain
  }) {
    super({
      config,
      state
    });
    _defineProperty(this, "name", "AddChainController");
    _defineProperty(this, "addChain", void 0);
    this.addChain = addChain;
    this.initialize();
  }
  async processAddChain(messageId) {
    const msgObject = this.getMessage(messageId);
    if (!msgObject) {
      throw new Error("Message not found");
    }
    try {
      await this.approveMessage(messageId, msgObject.messageParams);
      await this.addChain(msgObject.messageParams);
      this.updateMessage(_objectSpread(_objectSpread({}, msgObject), {}, {
        rawSig: JSON.stringify(msgObject.messageParams)
      }));
      this.setMessageStatus(messageId, baseControllers.MessageStatus.SIGNED);
      return null;
    } catch (error) {
      log.error(error);
      msgObject.error = (error === null || error === void 0 ? void 0 : error.message) || (error === null || error === void 0 ? void 0 : error.toString());
      this.setMessageStatus(messageId, baseControllers.MessageStatus.FAILED);
    }
  }
  async addNewUnapprovedMessage(messageParams, req) {
    await this.addUnapprovedMessage(messageParams, req);
    return this.waitForFinishStatus(messageParams, this.name);
  }
  async addUnapprovedMessage(messageParams, req) {
    // set message params origin first to satisfy the eslint rule (origin won't be checked by validateAddChainData)
    // for "Possible race condition: `messageParams.origin` might be assigned based on an outdated state of `messageParams`"
    if (req) {
      messageParams.origin = req.origin;
    }
    await utils.validateAddChainData(messageParams);
    const messageId = messageParams.id || baseControllers.randomId();
    const messageData = {
      id: messageId,
      messageParams,
      status: baseControllers.MessageStatus.UNAPPROVED,
      time: Date.now(),
      type: constants.METHOD_TYPES.ADD_CHAIN
    };
    await this.addMessage(messageData);
    this.emit(baseControllers.MESSAGE_EVENTS.UNAPPROVED_MESSAGE, {
      messageData,
      req
    });
    return messageId;
  }
  prepMessageForSigning(messageParams) {
    return Promise.resolve(messageParams);
  }
}

exports.AddChainController = AddChainController;
