// @toruslabs/ethereum-controllers/dist/lib.cjs/index.js
'use strict';

var AccountTrackerController = require('./Account/AccountTrackerController.js');
var AccountAbstractionController = require('./AccountAbstraction/AccountAbstractionController.js');
var BiconomySmartAccount = require('./AccountAbstraction/smartAccounts/BiconomySmartAccount.js');
var KernelSmartAccount = require('./AccountAbstraction/smartAccounts/KernelSmartAccount.js');
var MetamaskSmartAccount = require('./AccountAbstraction/smartAccounts/MetamaskSmartAccount.js');
var NexusSmartAccount = require('./AccountAbstraction/smartAccounts/NexusSmartAccount.js');
var SafeSmartAccount = require('./AccountAbstraction/smartAccounts/SafeSmartAccount.js');
var TrustSmartAccount = require('./AccountAbstraction/smartAccounts/TrustSmartAccount.js');
var PollingBlockTracker = require('./Block/PollingBlockTracker.js');
var CurrencyController = require('./Currency/CurrencyController.js');
var GasFeeController = require('./Gas/GasFeeController.js');
var KeyringController = require('./Keyring/KeyringController.js');
var AddChainController = require('./Message/AddChainController.js');
var MessageController = require('./Message/MessageController.js');
var PersonalMessageController = require('./Message/PersonalMessageController.js');
var SwitchChainController = require('./Message/SwitchChainController.js');
var TypedMessageController = require('./Message/TypedMessageController.js');
var utils = require('./Message/utils.js');
var createEthereumMiddleware = require('./Network/createEthereumMiddleware.js');
var createJsonRpcClient = require('./Network/createJsonRpcClient.js');
var NetworkController = require('./Network/NetworkController.js');
var NftHandler = require('./Nfts/NftHandler.js');
var NftsController = require('./Nfts/NftsController.js');
var PreferencesController = require('./Preferences/PreferencesController.js');
var TokenHandler = require('./Tokens/TokenHandler.js');
var TokenRatesController = require('./Tokens/TokenRatesController.js');
var TokensController = require('./Tokens/TokensController.js');
var NonceTracker = require('./Transaction/NonceTracker.js');
var PendingTransactionTracker = require('./Transaction/PendingTransactionTracker.js');
var TransactionController = require('./Transaction/TransactionController.js');
var TransactionGasUtil = require('./Transaction/TransactionGasUtil.js');
var TransactionStateHistoryHelper = require('./Transaction/TransactionStateHistoryHelper.js');
var TransactionStateManager = require('./Transaction/TransactionStateManager.js');
var TransactionUtils = require('./Transaction/TransactionUtils.js');
var abis = require('./utils/abis.js');
var constants = require('./utils/constants.js');
var conversionUtils = require('./utils/conversionUtils.js');
var helpers = require('./utils/helpers.js');



exports.AccountTrackerController = AccountTrackerController.AccountTrackerController;
exports.AccountAbstractionController = AccountAbstractionController.AccountAbstractionController;
exports.eoaInterceptorMiddleware = AccountAbstractionController.eoaInterceptorMiddleware;
exports.eoaProviderAsMiddleware = AccountAbstractionController.eoaProviderAsMiddleware;
exports.BiconomySmartAccount = BiconomySmartAccount.BiconomySmartAccount;
exports.KernelSmartAccount = KernelSmartAccount.KernelSmartAccount;
exports.MetamaskSmartAccount = MetamaskSmartAccount.MetamaskSmartAccount;
exports.NexusSmartAccount = NexusSmartAccount.NexusSmartAccount;
exports.SafeSmartAccount = SafeSmartAccount.SafeSmartAccount;
exports.TrustSmartAccount = TrustSmartAccount.TrustSmartAccount;
exports.PollingBlockTracker = PollingBlockTracker.PollingBlockTracker;
exports.CurrencyController = CurrencyController.CurrencyController;
exports.GasFeeController = GasFeeController.GasFeeController;
exports.KeyringController = KeyringController.KeyringController;
exports.AddChainController = AddChainController.AddChainController;
exports.MessageController = MessageController.MessageController;
exports.PersonalMessageController = PersonalMessageController.PersonalMessageController;
exports.SwitchChainController = SwitchChainController.SwitchChainController;
exports.TypedMessageController = TypedMessageController.TypedMessageController;
exports.normalizeMessageData = utils.normalizeMessageData;
exports.validateAddChainData = utils.validateAddChainData;
exports.validateAddress = utils.validateAddress;
exports.validateSignMessageData = utils.validateSignMessageData;
exports.validateSwitchChainData = utils.validateSwitchChainData;
exports.validateTypedSignMessageDataV4 = utils.validateTypedSignMessageDataV4;
exports.createAAMiddleware = createEthereumMiddleware.createAAMiddleware;
exports.createEthereumMiddleware = createEthereumMiddleware.createEthereumMiddleware;
exports.createGetAccountsMiddleware = createEthereumMiddleware.createGetAccountsMiddleware;
exports.createPendingNonceMiddleware = createEthereumMiddleware.createPendingNonceMiddleware;
exports.createPendingTxMiddleware = createEthereumMiddleware.createPendingTxMiddleware;
exports.createProcessAddEthereumChain = createEthereumMiddleware.createProcessAddEthereumChain;
exports.createProcessEstimateUserOperationGasMiddleware = createEthereumMiddleware.createProcessEstimateUserOperationGasMiddleware;
exports.createProcessEthSignMessage = createEthereumMiddleware.createProcessEthSignMessage;
exports.createProcessPersonalMessage = createEthereumMiddleware.createProcessPersonalMessage;
exports.createProcessSignTransactionMiddleware = createEthereumMiddleware.createProcessSignTransactionMiddleware;
exports.createProcessSwitchEthereumChain = createEthereumMiddleware.createProcessSwitchEthereumChain;
exports.createProcessTransactionMiddleware = createEthereumMiddleware.createProcessTransactionMiddleware;
exports.createProcessTypedMessageV4 = createEthereumMiddleware.createProcessTypedMessageV4;
exports.createProcessWalletSwitchChain = createEthereumMiddleware.createProcessWalletSwitchChain;
exports.createRequestAccountsMiddleware = createEthereumMiddleware.createRequestAccountsMiddleware;
exports.formatTxMetaForRpcResult = createEthereumMiddleware.formatTxMetaForRpcResult;
exports.createChainIdMiddleware = createJsonRpcClient.createChainIdMiddleware;
exports.createJsonRpcClient = createJsonRpcClient.createJsonRpcClient;
exports.createProviderConfigMiddleware = createJsonRpcClient.createProviderConfigMiddleware;
exports.NetworkController = NetworkController.NetworkController;
exports.NftHandler = NftHandler.NftHandler;
exports.NftsController = NftsController.NftsController;
exports.PreferencesController = PreferencesController.PreferencesController;
exports.TokenHandler = TokenHandler.TokenHandler;
exports.DEFAULT_CURRENCY = TokenRatesController.DEFAULT_CURRENCY;
exports.TokenRatesController = TokenRatesController.TokenRatesController;
exports.TokensController = TokensController.TokensController;
exports.NonceTracker = NonceTracker.NonceTracker;
exports.PendingTransactionTracker = PendingTransactionTracker.PendingTransactionTracker;
exports.TransactionController = TransactionController.TransactionController;
exports.TransactionGasUtil = TransactionGasUtil.TransactionGasUtil;
exports.generateHistoryEntry = TransactionStateHistoryHelper.generateHistoryEntry;
exports.replayHistory = TransactionStateHistoryHelper.replayHistory;
exports.snapshotFromTxMeta = TransactionStateHistoryHelper.snapshotFromTxMeta;
exports.TransactionStateManager = TransactionStateManager.TransactionStateManager;
exports.determineTransactionType = TransactionUtils.determineTransactionType;
exports.ensureFieldIsString = TransactionUtils.ensureFieldIsString;
exports.ensureMutuallyExclusiveFieldsNotProvided = TransactionUtils.ensureMutuallyExclusiveFieldsNotProvided;
exports.getFinalStates = TransactionUtils.getFinalStates;
exports.isEIP1559Transaction = TransactionUtils.isEIP1559Transaction;
exports.isLegacyTransaction = TransactionUtils.isLegacyTransaction;
exports.normalizeAndValidateTxParams = TransactionUtils.normalizeAndValidateTxParams;
exports.normalizeTxParameters = TransactionUtils.normalizeTxParameters;
exports.parseStandardTokenTransactionData = TransactionUtils.parseStandardTokenTransactionData;
exports.readAddressAsContract = TransactionUtils.readAddressAsContract;
exports.transactionMatchesNetwork = TransactionUtils.transactionMatchesNetwork;
exports.validateFrom = TransactionUtils.validateFrom;
exports.validateRecipient = TransactionUtils.validateRecipient;
exports.validateTxParameters = TransactionUtils.validateTxParameters;
exports.erc1155Abi = abis.erc1155Abi;
exports.erc20Abi = abis.erc20Abi;
exports.erc721Abi = abis.erc721Abi;
exports.singleBalanceCheckerAbi = abis.singleBalanceCheckerAbi;
exports.ARBITRUM_MAINNET_CHAIN_ID = constants.ARBITRUM_MAINNET_CHAIN_ID;
exports.ARBITRUM_TESTNET_CHAIN_ID = constants.ARBITRUM_TESTNET_CHAIN_ID;
exports.AVALANCHE_MAINNET_CHAIN_ID = constants.AVALANCHE_MAINNET_CHAIN_ID;
exports.AVALANCHE_TESTNET_CHAIN_ID = constants.AVALANCHE_TESTNET_CHAIN_ID;
exports.BASE_CHAIN_ID = constants.BASE_CHAIN_ID;
exports.BASE_TESTNET_CHAIN_ID = constants.BASE_TESTNET_CHAIN_ID;
exports.BSC_MAINNET_CHAIN_ID = constants.BSC_MAINNET_CHAIN_ID;
exports.BSC_TESTNET_CHAIN_ID = constants.BSC_TESTNET_CHAIN_ID;
exports.BUNDLER_METHOD_TYPES = constants.BUNDLER_METHOD_TYPES;
exports.CELO_MAINNET_CHAIN_ID = constants.CELO_MAINNET_CHAIN_ID;
exports.CHAIN_ID_TO_GAS_LIMIT_BUFFER_MAP = constants.CHAIN_ID_TO_GAS_LIMIT_BUFFER_MAP;
exports.COINGECKO_PLATFORMS_CHAIN_CODE_MAP = constants.COINGECKO_PLATFORMS_CHAIN_CODE_MAP;
exports.COINGECKO_SUPPORTED_CURRENCIES = constants.COINGECKO_SUPPORTED_CURRENCIES;
exports.CONTRACT_TYPE_ERC1155 = constants.CONTRACT_TYPE_ERC1155;
exports.CONTRACT_TYPE_ERC20 = constants.CONTRACT_TYPE_ERC20;
exports.CONTRACT_TYPE_ERC721 = constants.CONTRACT_TYPE_ERC721;
exports.CONTRACT_TYPE_ETH = constants.CONTRACT_TYPE_ETH;
exports.ERC1155_INTERFACE_ID = constants.ERC1155_INTERFACE_ID;
exports.ERC721_ENUMERABLE_INTERFACE_ID = constants.ERC721_ENUMERABLE_INTERFACE_ID;
exports.ERC721_INTERFACE_ID = constants.ERC721_INTERFACE_ID;
exports.ERC721_METADATA_INTERFACE_ID = constants.ERC721_METADATA_INTERFACE_ID;
exports.GAS_ESTIMATE_TYPES = constants.GAS_ESTIMATE_TYPES;
exports.LINEA_CHAIN_ID = constants.LINEA_CHAIN_ID;
exports.LINEA_SEPOLIA_CHAIN_ID = constants.LINEA_SEPOLIA_CHAIN_ID;
exports.LOCALHOST = constants.LOCALHOST;
exports.MAINNET_CHAIN_ID = constants.MAINNET_CHAIN_ID;
exports.METHOD_TYPES = constants.METHOD_TYPES;
exports.MM_NFT_API_SUPPORTED_CHAINS = constants.MM_NFT_API_SUPPORTED_CHAINS;
exports.MM_TOKEN_API_SUPPORTED_CHAINS = constants.MM_TOKEN_API_SUPPORTED_CHAINS;
exports.OLD_ERC721_LIST = constants.OLD_ERC721_LIST;
exports.OPTIMISM_MAINNET_CHAIN_ID = constants.OPTIMISM_MAINNET_CHAIN_ID;
exports.OPTIMISM_TESTNET_CHAIN_ID = constants.OPTIMISM_TESTNET_CHAIN_ID;
exports.PAYMASTER_METHOD_TYPES = constants.PAYMASTER_METHOD_TYPES;
exports.POLYGON_AMOY_CHAIN_ID = constants.POLYGON_AMOY_CHAIN_ID;
exports.POLYGON_CHAIN_ID = constants.POLYGON_CHAIN_ID;
exports.SEPOLIA_CHAIN_ID = constants.SEPOLIA_CHAIN_ID;
exports.SMART_ACCOUNT = constants.SMART_ACCOUNT;
exports.SUPPORTED_NETWORKS = constants.SUPPORTED_NETWORKS;
exports.TEST_CHAINS = constants.TEST_CHAINS;
exports.TRANSACTION_ENVELOPE_TYPES = constants.TRANSACTION_ENVELOPE_TYPES;
exports.XDAI_CHAIN_ID = constants.XDAI_CHAIN_ID;
exports.addCurrencies = conversionUtils.addCurrencies;
exports.conversionGTE = conversionUtils.conversionGTE;
exports.conversionGreaterThan = conversionUtils.conversionGreaterThan;
exports.conversionLTE = conversionUtils.conversionLTE;
exports.conversionLessThan = conversionUtils.conversionLessThan;
exports.conversionMax = conversionUtils.conversionMax;
exports.conversionUtil = conversionUtils.conversionUtil;
exports.decGWEIToHexWEI = conversionUtils.decGWEIToHexWEI;
exports.getBigNumber = conversionUtils.getBigNumber;
exports.hexWEIToDecGWEI = conversionUtils.hexWEIToDecGWEI;
exports.multiplyCurrencies = conversionUtils.multiplyCurrencies;
exports.subtractCurrencies = conversionUtils.subtractCurrencies;
exports.toNegative = conversionUtils.toNegative;
exports.BNToHex = helpers.BNToHex;
exports.GAS_LIMITS = helpers.GAS_LIMITS;
exports.addEtherscanTransactions = helpers.addEtherscanTransactions;
exports.bnLessThan = helpers.bnLessThan;
exports.formatDate = helpers.formatDate;
exports.formatPastTx = helpers.formatPastTx;
exports.formatTime = helpers.formatTime;
exports.getChainType = helpers.getChainType;
exports.getEthTxStatus = helpers.getEthTxStatus;
exports.getEtherScanHashLink = helpers.getEtherScanHashLink;
exports.getIpfsEndpoint = helpers.getIpfsEndpoint;
exports.hexToBn = helpers.hexToBn;
exports.isAddressByChainId = helpers.isAddressByChainId;
exports.sanitizeNftMetdataUrl = helpers.sanitizeNftMetdataUrl;
exports.toChecksumAddressByChainId = helpers.toChecksumAddressByChainId;
