// @toruslabs/base-controllers/dist/lib.esm/Popup/BroadcastChannelHandler.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { BroadcastChannel } from '@toruslabs/broadcast-channel';
import { POPUP_LOADED } from '../enums.js';
import { broadcastChannelOptions } from '../utils/utils.js';

class BroadcastChannelHandler {
  constructor(channelPrefix, instanceId) {
    _defineProperty(this, "bc", void 0);
    _defineProperty(this, "channel", void 0);
    const queryParameters = new URLSearchParams(window.location.search);
    const windowId = queryParameters.get("windowId");
    this.channel = `${channelPrefix}_${instanceId}_${windowId}`;
    this.bc = new BroadcastChannel(this.channel, broadcastChannelOptions);
  }
  getMessageFromChannel() {
    return new Promise((resolve, reject) => {
      this.bc.addEventListener("message", async ev => {
        this.bc.close();
        if (ev.error) {
          reject(ev.error);
        } else {
          resolve(ev.data);
        }
      });
      this.bc.postMessage({
        data: {
          type: POPUP_LOADED
        }
      });
    });
  }
}

export { BroadcastChannelHandler };
