// @toruslabs/customauth/dist/lib.cjs/utils/enums.js
'use strict';

const AUTH_CONNECTION = {
  // start - byoa
  GOOGLE: "google",
  TWITTER: "twitter",
  FACEBOOK: "facebook",
  DISCORD: "discord",
  FARCASTER: "farcaster",
  APPLE: "apple",
  GITHUB: "github",
  REDDIT: "reddit",
  LINE: "line",
  KAKAO: "kakao",
  LINKEDIN: "linkedin",
  TWITCH: "twitch",
  TELEGRAM: "telegram",
  WECHAT: "wechat",
  EMAIL_PASSWORDLESS: "email_passwordless",
  SMS_PASSWORDLESS: "sms_passwordless",
  // end - byoa
  CUSTOM: "custom",
  PASSKEYS: "passkeys",
  AUTHENTICATOR: "authenticator"
};
const UX_MODE = {
  POPUP: "popup",
  REDIRECT: "redirect"
};
const REDIRECT_PARAMS_STORAGE_METHOD = {
  LOCAL_STORAGE: "localStorage",
  SESSION_STORAGE: "sessionStorage",
  SERVER: "server"
};
const SENTRY_TXNS = {
  FETCH_NODE_DETAILS: "fetchNodeDetails",
  PUB_ADDRESS_LOOKUP: "pubAddressLookup",
  FETCH_SHARES: "fetchShares"
};

exports.AUTH_CONNECTION = AUTH_CONNECTION;
exports.REDIRECT_PARAMS_STORAGE_METHOD = REDIRECT_PARAMS_STORAGE_METHOD;
exports.SENTRY_TXNS = SENTRY_TXNS;
exports.UX_MODE = UX_MODE;
